{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "reverse-alberta",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "regulation-attitude",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from time import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "monthly-image",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cProfile\n",
    "import pstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "mobile-honduras",
   "metadata": {},
   "outputs": [],
   "source": [
    "%aimport malariagen_data.ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "judicial-raise",
   "metadata": {},
   "outputs": [],
   "source": [
    "%aimport malariagen_data.veff "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "living-wrestling",
   "metadata": {},
   "outputs": [],
   "source": [
    "veff = malariagen_data.veff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "younger-diagnosis",
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(\"simplecache::gs://vo_agam_release\", simplecache=dict(cache_storage=\"gcs_cache\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "wound-chassis",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<zarr.hierarchy.Group '/'>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome = ag3.open_genome()\n",
    "genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "composite-gallery",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['2L', '2R', '3L', '3R', 'Mt', 'UNKN', 'X', 'Y_unplaced']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "perceived-found",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCTCTACGTT'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome[\"3R\"][0:10].tobytes().decode()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "terminal-auditor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>49364325</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>157348</td>\n",
       "      <td>186936</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>157348</td>\n",
       "      <td>181305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677-RA</td>\n",
       "      <td>AGAP004677</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>157348</td>\n",
       "      <td>157495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2L</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>157348</td>\n",
       "      <td>157623</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP004677-RA</td>\n",
       "      <td>AGAP004677-RB-E4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196140</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>47932</td>\n",
       "      <td>48111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196141</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>47932</td>\n",
       "      <td>48138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>AGAP029375-RA-E2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196142</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>48112</td>\n",
       "      <td>48138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGAP029375-PA</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196143</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>exon</td>\n",
       "      <td>48301</td>\n",
       "      <td>48385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>AGAP029375-RA-E3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196144</th>\n",
       "      <td>Y_unplaced</td>\n",
       "      <td>VectorBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>48301</td>\n",
       "      <td>48385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AGAP029375-PA</td>\n",
       "      <td>AGAP029375-RA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196145 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid      source             type   start       end  score  \\\n",
       "0               2L  VectorBase       chromosome       1  49364325    NaN   \n",
       "1               2L  VectorBase             gene  157348    186936    NaN   \n",
       "2               2L  VectorBase             mRNA  157348    181305    NaN   \n",
       "3               2L  VectorBase  three_prime_UTR  157348    157495    NaN   \n",
       "4               2L  VectorBase             exon  157348    157623    NaN   \n",
       "...            ...         ...              ...     ...       ...    ...   \n",
       "196140  Y_unplaced  VectorBase   five_prime_UTR   47932     48111    NaN   \n",
       "196141  Y_unplaced  VectorBase             exon   47932     48138    NaN   \n",
       "196142  Y_unplaced  VectorBase              CDS   48112     48138    NaN   \n",
       "196143  Y_unplaced  VectorBase             exon   48301     48385    NaN   \n",
       "196144  Y_unplaced  VectorBase              CDS   48301     48385    NaN   \n",
       "\n",
       "       strand  phase             ID         Parent              Name  \n",
       "0         NaN    NaN             2L            NaN               NaN  \n",
       "1           -    NaN     AGAP004677            NaN               NaN  \n",
       "2           -    NaN  AGAP004677-RA     AGAP004677               NaN  \n",
       "3           -    NaN            NaN  AGAP004677-RA               NaN  \n",
       "4           -    NaN            NaN  AGAP004677-RA  AGAP004677-RB-E4  \n",
       "...       ...    ...            ...            ...               ...  \n",
       "196140      +    NaN            NaN  AGAP029375-RA               NaN  \n",
       "196141      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E2  \n",
       "196142      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN  \n",
       "196143      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E3  \n",
       "196144      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN  \n",
       "\n",
       "[196145 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneset = ag3.geneset()\n",
    "geneset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bibliographic-partnership",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 18.9 s, sys: 51.5 ms, total: 19 s\n",
      "Wall time: 19 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# this will take some time, loading gff, one-off cost\n",
    "ann = veff.Annotator(\n",
    "    genome=genome,\n",
    "    geneset=geneset,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "trained-supplement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VariantEffect(effect='NON_SYNONYMOUS_CODING', impact='MODERATE', chrom='2L', pos=2422652, ref='A', alt='T', vlen=0, ref_start=2422652, ref_stop=2422652, gene_id='AGAP004707', gene_start=2358158, gene_stop=2431617, gene_strand='+', transcript_id='AGAP004707-RD', transcript_start=2358158, transcript_stop=2431617, transcript_strand='+', cds_id='AGAP004707-PD', cds_start=2422468, cds_stop=2422655, cds_strand='+', intron_start=None, intron_stop=None, intron_5prime_dist=None, intron_3prime_dist=None, intron_cds_5prime=None, intron_cds_3prime=None, ref_cds_start=2984, ref_cds_stop=2984, ref_intron_start=None, ref_intron_stop=None, ref_start_phase=2, ref_codon='ttA', alt_codon='ttT', codon_change='ttA/ttT', aa_pos=995, ref_aa='L', alt_aa='F', aa_change='L995F')\n",
      "CPU times: user 0 ns, sys: 1.15 ms, total: 1.15 ms\n",
      "Wall time: 1.09 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "for effect in ann.get_effects(chrom='2L', pos=2422652, ref='A', alt='T',\n",
    "                                  transcript_ids=[\"AGAP004707-RD\"]):\n",
    "       print(effect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "great-region",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NON_SYNONYMOUS_CODING\n",
      "MODERATE\n",
      "2L\n",
      "2422652\n",
      "A\n",
      "T\n",
      "0\n",
      "2422652\n",
      "2422652\n",
      "AGAP004707\n",
      "2358158\n",
      "2431617\n",
      "+\n",
      "AGAP004707-RD\n",
      "2358158\n",
      "2431617\n",
      "+\n",
      "AGAP004707-PD\n",
      "2422468\n",
      "2422655\n",
      "+\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "None\n",
      "2984\n",
      "2984\n",
      "None\n",
      "None\n",
      "2\n",
      "ttA\n",
      "ttT\n",
      "ttA/ttT\n",
      "995\n",
      "L\n",
      "F\n",
      "L995F\n"
     ]
    }
   ],
   "source": [
    "for a in effect:\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "commercial-column",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('NON_SYNONYMOUS_CODING', 'MODERATE')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect.effect, effect.impact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "existing-hotel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ttA', 'ttT')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect.ref_codon, effect.alt_codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "rotary-values",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(995, 'L', 'F', 'L995F')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "effect.aa_pos, effect.ref_aa, effect.alt_aa, effect.aa_change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "northern-massage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "int"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(effect.aa_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "warming-retrieval",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos, ref, alt = ag3.snp_sites('2L')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "acting-commitment",
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "level-hazard",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2150929"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_idx = allel.SortedIndex(pos).locate_key(2429745)\n",
    "start_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "approved-jackson",
   "metadata": {},
   "outputs": [],
   "source": [
    "pp = pos[start_idx:start_idx+100].compute()\n",
    "rr = ref[start_idx:start_idx+100].compute()\n",
    "aaa = alt[start_idx:start_idx+100].compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "deluxe-festival",
   "metadata": {},
   "outputs": [],
   "source": [
    "def testf(n, show=False):\n",
    "    # loop over sites\n",
    "    for i, (p, r, aa) in enumerate(zip(pp, rr, aaa)):\n",
    "        if i < n:\n",
    "            # loop over alt alleles\n",
    "            for a in aa:\n",
    "                for effect in ann.get_effects(chrom='2L', pos=p, ref=r.decode(), alt=a.decode(),\n",
    "                                              transcript_ids=[\"AGAP004707-RA\"]):\n",
    "                    if show:\n",
    "                        print(effect)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "extra-hunger",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 50.7 ms, sys: 88 µs, total: 50.8 ms\n",
      "Wall time: 49.7 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "testf(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "patient-doubt",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.23 s, sys: 168 ms, total: 1.4 s\n",
      "Wall time: 4.12 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "x = ref.compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "honest-poultry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "48525747"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.nbytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "instructional-gibraltar",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         468523 function calls (464545 primitive calls) in 0.239 seconds\n",
      "\n",
      "   Ordered by: cumulative time\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "        1    0.000    0.000    0.239    0.239 {built-in method builtins.exec}\n",
      "        1    0.000    0.000    0.239    0.239 <string>:1(<module>)\n",
      "        1    0.001    0.001    0.239    0.239 <ipython-input-33-ac85f549eef5>:1(testf)\n",
      "      600    0.001    0.000    0.238    0.000 veff.py:134(get_effects)\n",
      "      600    0.002    0.000    0.139    0.000 veff.py:219(_get_gene_effects)\n",
      "      600    0.005    0.000    0.128    0.000 veff.py:257(_get_transcript_effects)\n",
      "      600    0.002    0.000    0.120    0.000 veff.py:324(_get_within_transcript_effects)\n",
      "      300    0.000    0.000    0.085    0.000 veff.py:57(find)\n",
      "      300    0.000    0.000    0.085    0.000 intervals.py:216(search)\n",
      "      300    0.001    0.000    0.084    0.000 intervals.py:190(_search_tree)\n",
      "      300    0.001    0.000    0.062    0.000 intervaltree.py:837(overlap)\n",
      " 4284/306    0.022    0.000    0.058    0.000 node.py:309(search_point)\n",
      "   105747    0.042    0.000    0.053    0.000 base.py:570(__getattr__)\n",
      "      171    0.002    0.000    0.037    0.000 veff.py:405(_get_cds_effect)\n",
      "     9036    0.012    0.000    0.036    0.000 {method 'add' of 'set' objects}\n",
      "      771    0.008    0.000    0.035    0.000 {built-in method builtins.sorted}\n",
      "      171    0.001    0.000    0.033    0.000 veff.py:450(_get_within_cds_effect)\n",
      "      171    0.001    0.000    0.031    0.000 veff.py:602(_get_aa_change)\n",
      "      300    0.017    0.000    0.026    0.000 veff.py:355(<listcomp>)\n",
      "      171    0.002    0.000    0.023    0.000 veff.py:630(_get_codon_change)\n",
      "    84273    0.021    0.000    0.021    0.000 interval.py:159(__eq__)\n",
      "      171    0.002    0.000    0.019    0.000 veff.py:707(_get_coding_position)\n",
      "    33030    0.005    0.000    0.016    0.000 interval.py:204(__lt__)\n",
      "      300    0.006    0.000    0.014    0.000 veff.py:341(<listcomp>)\n",
      "      300    0.006    0.000    0.014    0.000 veff.py:333(<listcomp>)\n",
      "   111048    0.012    0.000    0.012    0.000 {method 'index' of 'list' objects}\n",
      "    33030    0.011    0.000    0.011    0.000 interval.py:173(__cmp__)\n",
      "      129    0.001    0.000    0.009    0.000 veff.py:748(_get_intron_effect)\n",
      "      300    0.009    0.000    0.009    0.000 veff.py:359(<listcomp>)\n",
      "      129    0.001    0.000    0.008    0.000 veff.py:784(_get_within_intron_effect)\n",
      "      300    0.003    0.000    0.008    0.000 veff.py:186(<listcomp>)\n",
      "     9000    0.003    0.000    0.008    0.000 veff.py:337(<lambda>)\n",
      "      171    0.003    0.000    0.008    0.000 veff.py:720(<listcomp>)\n",
      "      342    0.002    0.000    0.007    0.000 Seq.py:918(translate)\n",
      "     1677    0.001    0.000    0.006    0.000 <string>:24(_replace)\n",
      "      342    0.002    0.000    0.005    0.000 Seq.py:2223(_translate_str)\n",
      "     1677    0.001    0.000    0.005    0.000 <string>:16(_make)\n",
      "      171    0.002    0.000    0.004    0.000 veff.py:716(<listcomp>)\n",
      "      471    0.002    0.000    0.004    0.000 veff.py:71(get_ref_allele_coords)\n",
      "     1677    0.004    0.000    0.004    0.000 {built-in method __new__ of type object at 0x9d12c0}\n",
      "     5130    0.003    0.000    0.004    0.000 base.py:557(__getitem__)\n",
      "      300    0.001    0.000    0.004    0.000 veff.py:248(<listcomp>)\n",
      "      300    0.001    0.000    0.003    0.000 veff.py:233(<listcomp>)\n",
      "      129    0.001    0.000    0.003    0.000 veff.py:798(<listcomp>)\n",
      "      129    0.001    0.000    0.003    0.000 veff.py:801(<listcomp>)\n",
      "     9036    0.002    0.000    0.003    0.000 interval.py:151(__hash__)\n",
      "      342    0.001    0.000    0.003    0.000 Seq.py:2300(<listcomp>)\n",
      "      870    0.002    0.000    0.003    0.000 veff.py:60(get_ref_seq)\n",
      "      600    0.001    0.000    0.003    0.000 sortedlist.py:1166(bisect_left)\n",
      "     3420    0.001    0.000    0.002    0.000 CodonTable.py:400(__contains__)\n",
      "     1200    0.002    0.000    0.002    0.000 {built-in method _bisect.bisect_left}\n",
      "     3762    0.001    0.000    0.001    0.000 CodonTable.py:418(__getitem__)\n",
      "     8262    0.001    0.000    0.001    0.000 node.py:505(__getitem__)\n",
      "      300    0.000    0.000    0.001    0.000 node.py:300(search_overlap)\n",
      "     9036    0.001    0.000    0.001    0.000 {built-in method builtins.hash}\n",
      "     6504    0.001    0.000    0.001    0.000 {built-in method builtins.isinstance}\n",
      "      300    0.001    0.000    0.001    0.000 intervals.py:218(<listcomp>)\n",
      "     6450    0.000    0.000    0.000    0.000 {built-in method builtins.len}\n",
      "      684    0.000    0.000    0.000    0.000 Seq.py:74(__init__)\n",
      "      870    0.000    0.000    0.000    0.000 {method 'tobytes' of 'numpy.ndarray' objects}\n",
      "     1470    0.000    0.000    0.000    0.000 {method 'decode' of 'bytes' objects}\n",
      "      300    0.000    0.000    0.000    0.000 {method 'update' of 'set' objects}\n",
      "     1626    0.000    0.000    0.000    0.000 {method 'upper' of 'str' objects}\n",
      "      600    0.000    0.000    0.000    0.000 veff.py:54(get_children)\n",
      "     1341    0.000    0.000    0.000    0.000 {method 'lower' of 'str' objects}\n",
      "      684    0.000    0.000    0.000    0.000 Seq.py:111(__str__)\n",
      "      306    0.000    0.000    0.000    0.000 intervaltree.py:861(<genexpr>)\n",
      "      171    0.000    0.000    0.000    0.000 {built-in method builtins.sum}\n",
      "      600    0.000    0.000    0.000    0.000 sortedlist.py:520(_loc)\n",
      "      129    0.000    0.000    0.000    0.000 {built-in method builtins.min}\n",
      "      342    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}\n",
      "      342    0.000    0.000    0.000    0.000 {method 'join' of 'str' objects}\n",
      "        6    0.000    0.000    0.000    0.000 sorteddict.py:650(__getitem__)\n",
      "        6    0.000    0.000    0.000    0.000 sortedlist.py:816(__getitem__)\n",
      "        6    0.000    0.000    0.000    0.000 sorteddict.py:329(keys)\n",
      "        6    0.000    0.000    0.000    0.000 _collections_abc.py:698(__init__)\n",
      "        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cProfile.run('testf(100)', sort='cumtime')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "funny-waterproof",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         118564 function calls (117784 primitive calls) in 0.063 seconds\n",
      "\n",
      "   Ordered by: internal time\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "    22620    0.010    0.000    0.012    0.000 base.py:570(__getattr__)\n",
      "    25020    0.007    0.000    0.007    0.000 interval.py:159(__eq__)\n",
      "   840/60    0.005    0.000    0.017    0.000 node.py:309(search_point)\n",
      "     2460    0.004    0.000    0.012    0.000 {method 'add' of 'set' objects}\n",
      "       60    0.004    0.000    0.006    0.000 veff.py:355(<listcomp>)\n",
      "     9960    0.003    0.000    0.003    0.000 interval.py:173(__cmp__)\n",
      "    24480    0.003    0.000    0.003    0.000 {method 'index' of 'list' objects}\n",
      "      180    0.002    0.000    0.010    0.000 {built-in method builtins.sorted}\n",
      "       60    0.002    0.000    0.002    0.000 veff.py:359(<listcomp>)\n",
      "     9960    0.002    0.000    0.005    0.000 interval.py:204(__lt__)\n",
      "       60    0.001    0.000    0.003    0.000 veff.py:333(<listcomp>)\n",
      "       60    0.001    0.000    0.003    0.000 veff.py:341(<listcomp>)\n",
      "      120    0.001    0.000    0.031    0.000 veff.py:257(_get_transcript_effects)\n",
      "       60    0.001    0.000    0.003    0.000 veff.py:720(<listcomp>)\n",
      "     1800    0.001    0.000    0.001    0.000 base.py:557(__getitem__)\n",
      "       60    0.001    0.000    0.002    0.000 veff.py:186(<listcomp>)\n",
      "      360    0.001    0.000    0.001    0.000 {built-in method __new__ of type object at 0x9d12c0}\n",
      "      120    0.001    0.000    0.002    0.000 Seq.py:2223(_translate_str)\n",
      "      120    0.001    0.000    0.001    0.000 veff.py:71(get_ref_allele_coords)\n",
      "       60    0.001    0.000    0.013    0.000 veff.py:405(_get_cds_effect)\n",
      "     1800    0.001    0.000    0.002    0.000 veff.py:337(<lambda>)\n",
      "       60    0.001    0.000    0.008    0.000 veff.py:630(_get_codon_change)\n",
      "      264    0.001    0.000    0.001    0.000 veff.py:60(get_ref_seq)\n",
      "     2460    0.001    0.000    0.001    0.000 interval.py:151(__hash__)\n",
      "      120    0.001    0.000    0.003    0.000 Seq.py:918(translate)\n",
      "       60    0.001    0.000    0.002    0.000 veff.py:716(<listcomp>)\n",
      "     1200    0.001    0.000    0.001    0.000 CodonTable.py:400(__contains__)\n",
      "       60    0.001    0.000    0.007    0.000 veff.py:707(_get_coding_position)\n",
      "        1    0.000    0.000    0.063    0.063 <ipython-input-33-ac85f549eef5>:1(testf)\n",
      "      120    0.000    0.000    0.029    0.000 veff.py:324(_get_within_transcript_effects)\n",
      "       60    0.000    0.000    0.012    0.000 veff.py:450(_get_within_cds_effect)\n",
      "      360    0.000    0.000    0.002    0.000 <string>:24(_replace)\n",
      "     1320    0.000    0.000    0.000    0.000 CodonTable.py:418(__getitem__)\n",
      "      120    0.000    0.000    0.062    0.001 veff.py:134(get_effects)\n",
      "       60    0.000    0.000    0.025    0.000 intervals.py:190(_search_tree)\n",
      "       60    0.000    0.000    0.001    0.000 veff.py:233(<listcomp>)\n",
      "      240    0.000    0.000    0.000    0.000 {built-in method _bisect.bisect_left}\n",
      "       60    0.000    0.000    0.001    0.000 veff.py:248(<listcomp>)\n",
      "      120    0.000    0.000    0.033    0.000 veff.py:219(_get_gene_effects)\n",
      "       60    0.000    0.000    0.011    0.000 veff.py:602(_get_aa_change)\n",
      "      120    0.000    0.000    0.001    0.000 Seq.py:2300(<listcomp>)\n",
      "       60    0.000    0.000    0.018    0.000 intervaltree.py:837(overlap)\n",
      "     2280    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}\n",
      "      120    0.000    0.000    0.001    0.000 sortedlist.py:1166(bisect_left)\n",
      "      360    0.000    0.000    0.001    0.000 <string>:16(_make)\n",
      "     2460    0.000    0.000    0.000    0.000 {built-in method builtins.hash}\n",
      "       60    0.000    0.000    0.000    0.000 intervals.py:218(<listcomp>)\n",
      "      240    0.000    0.000    0.000    0.000 Seq.py:74(__init__)\n",
      "     1584    0.000    0.000    0.000    0.000 {built-in method builtins.len}\n",
      "     1620    0.000    0.000    0.000    0.000 node.py:505(__getitem__)\n",
      "       60    0.000    0.000    0.025    0.000 intervals.py:216(search)\n",
      "       60    0.000    0.000    0.000    0.000 node.py:300(search_overlap)\n",
      "      264    0.000    0.000    0.000    0.000 {method 'tobytes' of 'numpy.ndarray' objects}\n",
      "      120    0.000    0.000    0.000    0.000 veff.py:54(get_children)\n",
      "      384    0.000    0.000    0.000    0.000 {method 'decode' of 'bytes' objects}\n",
      "      384    0.000    0.000    0.000    0.000 {method 'lower' of 'str' objects}\n",
      "        1    0.000    0.000    0.063    0.063 {built-in method builtins.exec}\n",
      "      480    0.000    0.000    0.000    0.000 {method 'upper' of 'str' objects}\n",
      "       60    0.000    0.000    0.025    0.000 veff.py:57(find)\n",
      "      240    0.000    0.000    0.000    0.000 Seq.py:111(__str__)\n",
      "       60    0.000    0.000    0.000    0.000 {built-in method builtins.sum}\n",
      "      120    0.000    0.000    0.000    0.000 {method 'join' of 'str' objects}\n",
      "       60    0.000    0.000    0.000    0.000 intervaltree.py:861(<genexpr>)\n",
      "      120    0.000    0.000    0.000    0.000 sortedlist.py:520(_loc)\n",
      "      120    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}\n",
      "       60    0.000    0.000    0.000    0.000 {method 'update' of 'set' objects}\n",
      "        1    0.000    0.000    0.063    0.063 <string>:1(<module>)\n",
      "        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cProfile.run('testf(20)', sort='time')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "tight-strengthening",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('2L',\n",
       " 'VectorBase',\n",
       " 'mRNA',\n",
       " 2358158,\n",
       " 2431617,\n",
       " nan,\n",
       " '+',\n",
       " nan,\n",
       " 'AGAP004707-RD',\n",
       " 'AGAP004707',\n",
       " nan)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann.get_feature(\"AGAP004707-RD\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "alien-custom",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RA',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RB',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RC',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RD',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RE',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RF',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RG',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RH',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RI',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RJ',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RK',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RL',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RM',\n",
       "  'AGAP004707',\n",
       "  nan)]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann.get_children(\"AGAP004707\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "egyptian-hawaii",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('2L',\n",
       "  'VectorBase',\n",
       "  'chromosome',\n",
       "  1,\n",
       "  49364325,\n",
       "  nan,\n",
       "  nan,\n",
       "  nan,\n",
       "  '2L',\n",
       "  nan,\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PE',\n",
       "  'AGAP004707-RE',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PK',\n",
       "  'AGAP004707-RK',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PI',\n",
       "  'AGAP004707-RI',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PA',\n",
       "  'AGAP004707-RA',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PB',\n",
       "  'AGAP004707-RB',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PD',\n",
       "  'AGAP004707-RD',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PJ',\n",
       "  'AGAP004707-RJ',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PL',\n",
       "  'AGAP004707-RL',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PM',\n",
       "  'AGAP004707-RM',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PF',\n",
       "  'AGAP004707-RF',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PG',\n",
       "  'AGAP004707-RG',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PH',\n",
       "  'AGAP004707-RH',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'CDS',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  0.0,\n",
       "  'AGAP004707-PC',\n",
       "  'AGAP004707-RC',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RL',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RB',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RD',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RE',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RJ',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RK',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RM',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RF',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RG',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RH',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RI',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RA',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'exon',\n",
       "  2358158,\n",
       "  2358304,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  nan,\n",
       "  'AGAP004707-RC',\n",
       "  'AGAP004707-RA-E1'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'gene',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707',\n",
       "  nan,\n",
       "  'para'),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RJ',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RK',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RL',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RM',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RB',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RA',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RC',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RD',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RE',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RF',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RG',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RH',\n",
       "  'AGAP004707',\n",
       "  nan),\n",
       " ('2L',\n",
       "  'VectorBase',\n",
       "  'mRNA',\n",
       "  2358158,\n",
       "  2431617,\n",
       "  nan,\n",
       "  '+',\n",
       "  nan,\n",
       "  'AGAP004707-RI',\n",
       "  'AGAP004707',\n",
       "  nan)]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann.find(\"2L\", 2358158, 2358159)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "assumed-creature",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transcript : AGAP004707-RA\n",
      "chromosome : 2L \n",
      "start : 2358158\n",
      "stop : 2431617\n",
      "strand : +\n",
      "CPU times: user 46.8 s, sys: 486 ms, total: 47.3 s\n",
      "Wall time: 46.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "df = ag3.snp_effects('AGAP004707-RA', site_mask='gamb_colu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "northern-priority",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_alleles</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>ref_codon</th>\n",
       "      <th>alt_codon</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>aa_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2358158</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>START_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Atg</td>\n",
       "      <td>Ctg</td>\n",
       "      <td>1.0</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>M1L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2358158</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>START_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Atg</td>\n",
       "      <td>Ttg</td>\n",
       "      <td>1.0</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>M1L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2358158</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>START_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Atg</td>\n",
       "      <td>Gtg</td>\n",
       "      <td>1.0</td>\n",
       "      <td>M</td>\n",
       "      <td>V</td>\n",
       "      <td>M1V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2358159</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>aTg</td>\n",
       "      <td>aAg</td>\n",
       "      <td>1.0</td>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "      <td>M1K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2358159</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NON_SYNONYMOUS_CODING</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>aTg</td>\n",
       "      <td>aCg</td>\n",
       "      <td>1.0</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "      <td>M1T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132301</th>\n",
       "      <td>2431616</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>STOP_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>tGa</td>\n",
       "      <td>tCa</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>*</td>\n",
       "      <td>S</td>\n",
       "      <td>*2065S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132302</th>\n",
       "      <td>2431616</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>STOP_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>tGa</td>\n",
       "      <td>tTa</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>*</td>\n",
       "      <td>L</td>\n",
       "      <td>*2065L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132303</th>\n",
       "      <td>2431617</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>STOP_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>tgA</td>\n",
       "      <td>tgC</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>*</td>\n",
       "      <td>C</td>\n",
       "      <td>*2065C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132304</th>\n",
       "      <td>2431617</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>STOP_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>tgA</td>\n",
       "      <td>tgT</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>*</td>\n",
       "      <td>C</td>\n",
       "      <td>*2065C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132305</th>\n",
       "      <td>2431617</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>STOP_LOST</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>tgA</td>\n",
       "      <td>tgG</td>\n",
       "      <td>2065.0</td>\n",
       "      <td>*</td>\n",
       "      <td>W</td>\n",
       "      <td>*2065W</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132306 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        position ref_allele alt_alleles                 effect    impact  \\\n",
       "0        2358158          A           C             START_LOST      HIGH   \n",
       "1        2358158          A           T             START_LOST      HIGH   \n",
       "2        2358158          A           G             START_LOST      HIGH   \n",
       "3        2358159          T           A  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "4        2358159          T           C  NON_SYNONYMOUS_CODING  MODERATE   \n",
       "...          ...        ...         ...                    ...       ...   \n",
       "132301   2431616          G           C              STOP_LOST      HIGH   \n",
       "132302   2431616          G           T              STOP_LOST      HIGH   \n",
       "132303   2431617          A           C              STOP_LOST      HIGH   \n",
       "132304   2431617          A           T              STOP_LOST      HIGH   \n",
       "132305   2431617          A           G              STOP_LOST      HIGH   \n",
       "\n",
       "       ref_codon alt_codon  aa_pos ref_aa alt_aa aa_change  \n",
       "0            Atg       Ctg     1.0      M      L       M1L  \n",
       "1            Atg       Ttg     1.0      M      L       M1L  \n",
       "2            Atg       Gtg     1.0      M      V       M1V  \n",
       "3            aTg       aAg     1.0      M      K       M1K  \n",
       "4            aTg       aCg     1.0      M      T       M1T  \n",
       "...          ...       ...     ...    ...    ...       ...  \n",
       "132301       tGa       tCa  2065.0      *      S    *2065S  \n",
       "132302       tGa       tTa  2065.0      *      L    *2065L  \n",
       "132303       tgA       tgC  2065.0      *      C    *2065C  \n",
       "132304       tgA       tgT  2065.0      *      C    *2065C  \n",
       "132305       tgA       tgG  2065.0      *      W    *2065W  \n",
       "\n",
       "[132306 rows x 11 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-democrat",
   "metadata": {},
   "source": [
    "# SNP allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "gross-rating",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transcript : AGAP004707-RD\n",
      "chromosome : 2L \n",
      "start : 2358158\n",
      "stop : 2431617\n",
      "strand : +\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/chris/.cache/pypoetry/virtualenvs/malariagen-data-mfr76-fD-py3.6/lib/python3.6/site-packages/dask/array/routines.py:1247: PerformanceWarning: Slicing is producing a large chunk. To accept the large\n",
      "chunk and silence this warning, set the option\n",
      "    >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):\n",
      "    ...     array[indexer]\n",
      "\n",
      "To avoid creating the large chunks, set the option\n",
      "    >>> with dask.config.set(**{'array.slicing.split_large_chunks': True}):\n",
      "    ...     array[indexer]\n",
      "  a = a[tuple(condition if i == axis else slice(None) for i in range(a.ndim))]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 44.4 s, sys: 7.24 s, total: 51.7 s\n",
      "Wall time: 3min 10s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "df_meta, gt = ag3.snp_allele_frequencies(\"AGAP004707-RD\", \"gamb_colu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "caring-building",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "132306"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gt) *3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "direct-typing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>contributor</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>sex_call</th>\n",
       "      <th>sample_set</th>\n",
       "      <th>release</th>\n",
       "      <th>aim_fraction_colu</th>\n",
       "      <th>aim_fraction_arab</th>\n",
       "      <th>species_gambcolu_arabiensis</th>\n",
       "      <th>species_gambiae_coluzzii</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2494</th>\n",
       "      <td>AC0007-C</td>\n",
       "      <td>1_A2</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>2012</td>\n",
       "      <td>10</td>\n",
       "      <td>0.770</td>\n",
       "      <td>34.026</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.458</td>\n",
       "      <td>0.741</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>arabiensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>AC0008-C</td>\n",
       "      <td>1_A6</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>2012</td>\n",
       "      <td>10</td>\n",
       "      <td>0.770</td>\n",
       "      <td>34.026</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.456</td>\n",
       "      <td>0.740</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>arabiensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>AC0009-Cx</td>\n",
       "      <td>1_A7</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>2012</td>\n",
       "      <td>10</td>\n",
       "      <td>0.770</td>\n",
       "      <td>34.026</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.458</td>\n",
       "      <td>0.746</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>arabiensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>AC0010-C</td>\n",
       "      <td>1_A9</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>2012</td>\n",
       "      <td>10</td>\n",
       "      <td>0.770</td>\n",
       "      <td>34.026</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.743</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>arabiensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>AC0011-C</td>\n",
       "      <td>1_B5</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>2012</td>\n",
       "      <td>10</td>\n",
       "      <td>0.770</td>\n",
       "      <td>34.026</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.455</td>\n",
       "      <td>0.745</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>arabiensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2779</th>\n",
       "      <td>AC0295-C</td>\n",
       "      <td>K92</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>2012</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.751</td>\n",
       "      <td>29.701</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2780</th>\n",
       "      <td>AC0296-C</td>\n",
       "      <td>K93</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>2012</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.751</td>\n",
       "      <td>29.701</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.003</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2781</th>\n",
       "      <td>AC0297-C</td>\n",
       "      <td>K94</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>2012</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.751</td>\n",
       "      <td>29.701</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782</th>\n",
       "      <td>AC0298-C</td>\n",
       "      <td>K95</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>2012</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.751</td>\n",
       "      <td>29.701</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2783</th>\n",
       "      <td>AC0299-C</td>\n",
       "      <td>K96</td>\n",
       "      <td>Martin Donnelly</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>2012</td>\n",
       "      <td>11</td>\n",
       "      <td>-0.751</td>\n",
       "      <td>29.701</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-UG</td>\n",
       "      <td>v3</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>290 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id partner_sample_id      contributor country   location  year  \\\n",
       "2494   AC0007-C              1_A2  Martin Donnelly  Uganda  Nagongera  2012   \n",
       "2495   AC0008-C              1_A6  Martin Donnelly  Uganda  Nagongera  2012   \n",
       "2496  AC0009-Cx              1_A7  Martin Donnelly  Uganda  Nagongera  2012   \n",
       "2497   AC0010-C              1_A9  Martin Donnelly  Uganda  Nagongera  2012   \n",
       "2498   AC0011-C              1_B5  Martin Donnelly  Uganda  Nagongera  2012   \n",
       "...         ...               ...              ...     ...        ...   ...   \n",
       "2779   AC0295-C               K92  Martin Donnelly  Uganda     Kihihi  2012   \n",
       "2780   AC0296-C               K93  Martin Donnelly  Uganda     Kihihi  2012   \n",
       "2781   AC0297-C               K94  Martin Donnelly  Uganda     Kihihi  2012   \n",
       "2782   AC0298-C               K95  Martin Donnelly  Uganda     Kihihi  2012   \n",
       "2783   AC0299-C               K96  Martin Donnelly  Uganda     Kihihi  2012   \n",
       "\n",
       "      month  latitude  longitude sex_call  sample_set release  \\\n",
       "2494     10     0.770     34.026        F  AG1000G-UG      v3   \n",
       "2495     10     0.770     34.026        F  AG1000G-UG      v3   \n",
       "2496     10     0.770     34.026        F  AG1000G-UG      v3   \n",
       "2497     10     0.770     34.026        F  AG1000G-UG      v3   \n",
       "2498     10     0.770     34.026        F  AG1000G-UG      v3   \n",
       "...     ...       ...        ...      ...         ...     ...   \n",
       "2779     11    -0.751     29.701        F  AG1000G-UG      v3   \n",
       "2780     11    -0.751     29.701        F  AG1000G-UG      v3   \n",
       "2781     11    -0.751     29.701        F  AG1000G-UG      v3   \n",
       "2782     11    -0.751     29.701        F  AG1000G-UG      v3   \n",
       "2783     11    -0.751     29.701        F  AG1000G-UG      v3   \n",
       "\n",
       "      aim_fraction_colu  aim_fraction_arab species_gambcolu_arabiensis  \\\n",
       "2494              0.458              0.741                  arabiensis   \n",
       "2495              0.456              0.740                  arabiensis   \n",
       "2496              0.458              0.746                  arabiensis   \n",
       "2497              0.452              0.743                  arabiensis   \n",
       "2498              0.455              0.745                  arabiensis   \n",
       "...                 ...                ...                         ...   \n",
       "2779              0.026              0.002                   gamb_colu   \n",
       "2780              0.029              0.003                   gamb_colu   \n",
       "2781              0.026              0.002                   gamb_colu   \n",
       "2782              0.029              0.002                   gamb_colu   \n",
       "2783              0.029              0.002                   gamb_colu   \n",
       "\n",
       "     species_gambiae_coluzzii     species  \n",
       "2494                      NaN  arabiensis  \n",
       "2495                      NaN  arabiensis  \n",
       "2496                      NaN  arabiensis  \n",
       "2497                      NaN  arabiensis  \n",
       "2498                      NaN  arabiensis  \n",
       "...                       ...         ...  \n",
       "2779                  gambiae     gambiae  \n",
       "2780                  gambiae     gambiae  \n",
       "2781                  gambiae     gambiae  \n",
       "2782                  gambiae     gambiae  \n",
       "2783                  gambiae     gambiae  \n",
       "\n",
       "[290 rows x 17 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta.loc[df_meta.country == 'AG1000G-UG']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "competitive-commons",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
