
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Ag3 &#8212; malariagen_data API  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=5929fcd5"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Ag3';</script>
    <script>
        DOCUMENTATION_OPTIONS.theme_version = '0.16.1';
        DOCUMENTATION_OPTIONS.theme_switcher_json_url = 'https://malariagen.github.io/malariagen-data-python/latest/_static/switcher.json';
        DOCUMENTATION_OPTIONS.theme_switcher_version_match = 'dev';
        DOCUMENTATION_OPTIONS.show_version_warning_banner =
            false;
        </script>
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="dev" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.svg" class="logo__image only-light" alt="malariagen_data API  documentation - Home"/>
    <img src="_static/logo.svg" class="logo__image only-dark pst-js-only" alt="malariagen_data API  documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malariagen/malariagen-data-python" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-3"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-3"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-3"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-3">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
          
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/malariagen/malariagen-data-python" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Ag3</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="ag3">
<h1>Ag3<a class="headerlink" href="#ag3" title="Link to this heading">#</a></h1>
<p>This page provides a curated list of functions and properties available in the <code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code> API
for data on mosquitoes from the <em>Anopheles gambiae</em> complex.</p>
<p>To set up the API, use the following code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">malariagen_data</span>
<span class="n">ag3</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Ag3</span><span class="p">()</span>
</pre></div>
</div>
<p>All the functions below can then be accessed as methods on the <code class="docutils literal notranslate"><span class="pre">ag3</span></code> object. E.g., to call the
<code class="docutils literal notranslate"><span class="pre">sample_metadata()</span></code> function, do:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">()</span>
</pre></div>
</div>
<p>For more information about the data and terms of use, please see the
<a class="reference external" href="https://www.malariagen.net/project/anopheles-gambiae-genomic-surveillance-project">MalariaGEN Anopheles gambiae genomic surveillance project</a>
home page.</p>
<section id="basic-data-access">
<h2>Basic data access<a class="headerlink" href="#basic-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.releases.html#malariagen_data.ag3.Ag3.releases" title="malariagen_data.ag3.Ag3.releases"><code class="xref py py-obj docutils literal notranslate"><span class="pre">releases</span></code></a></p></td>
<td><p>Relevant data releases.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.sample_sets.html#malariagen_data.ag3.Ag3.sample_sets" title="malariagen_data.ag3.Ag3.sample_sets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_sets</span></code></a>([release])</p></td>
<td><p>Access a dataframe of sample sets.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.lookup_release.html#malariagen_data.ag3.Ag3.lookup_release" title="malariagen_data.ag3.Ag3.lookup_release"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lookup_release</span></code></a>(sample_set)</p></td>
<td><p>Find which release a sample set was included in.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.lookup_study.html#malariagen_data.ag3.Ag3.lookup_study" title="malariagen_data.ag3.Ag3.lookup_study"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lookup_study</span></code></a>(sample_set)</p></td>
<td><p>Find which study a sample set belongs to.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="reference-genome-data-access">
<h2>Reference genome data access<a class="headerlink" href="#reference-genome-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.contigs.html#malariagen_data.ag3.Ag3.contigs" title="malariagen_data.ag3.Ag3.contigs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">contigs</span></code></a></p></td>
<td><p>Contigs in the reference genome.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.genome_sequence.html#malariagen_data.ag3.Ag3.genome_sequence" title="malariagen_data.ag3.Ag3.genome_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genome_sequence</span></code></a>(region[, inline_array, chunks])</p></td>
<td><p>Access the reference genome sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.genome_features.html#malariagen_data.ag3.Ag3.genome_features" title="malariagen_data.ag3.Ag3.genome_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">genome_features</span></code></a>([region, attributes])</p></td>
<td><p>Access genome feature annotations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_transcript.html#malariagen_data.ag3.Ag3.plot_transcript" title="malariagen_data.ag3.Ag3.plot_transcript"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_transcript</span></code></a>(transcript[, sizing_mode, ...])</p></td>
<td><p>Plot a transcript, using bokeh.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_genes.html#malariagen_data.ag3.Ag3.plot_genes" title="malariagen_data.ag3.Ag3.plot_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_genes</span></code></a>(region[, sizing_mode, width, ...])</p></td>
<td><p>Plot a genes track, using bokeh.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="sample-metadata-access">
<h2>Sample metadata access<a class="headerlink" href="#sample-metadata-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.sample_metadata.html#malariagen_data.ag3.Ag3.sample_metadata" title="malariagen_data.ag3.Ag3.sample_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_metadata</span></code></a>([sample_sets, sample_query, ...])</p></td>
<td><p>Access sample metadata for one or more sample sets.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.add_extra_metadata.html#malariagen_data.ag3.Ag3.add_extra_metadata" title="malariagen_data.ag3.Ag3.add_extra_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_extra_metadata</span></code></a>(data[, on])</p></td>
<td><p>Add extra sample metadata, e.g., including additional columns which you would like to use to query and group samples.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.clear_extra_metadata.html#malariagen_data.ag3.Ag3.clear_extra_metadata" title="malariagen_data.ag3.Ag3.clear_extra_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear_extra_metadata</span></code></a>()</p></td>
<td><p>Clear any extra metadata previously added.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.cross_metadata.html#malariagen_data.ag3.Ag3.cross_metadata" title="malariagen_data.ag3.Ag3.cross_metadata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cross_metadata</span></code></a>()</p></td>
<td><p>Load a dataframe containing metadata about samples in colony crosses, including which samples are parents or progeny in which crosses.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.count_samples.html#malariagen_data.ag3.Ag3.count_samples" title="malariagen_data.ag3.Ag3.count_samples"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_samples</span></code></a>([sample_sets, sample_query, ...])</p></td>
<td><p>Create a pivot table showing numbers of samples available by space, time and taxon.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.lookup_sample.html#malariagen_data.ag3.Ag3.lookup_sample" title="malariagen_data.ag3.Ag3.lookup_sample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lookup_sample</span></code></a>(sample[, sample_set])</p></td>
<td><p>Get the metadata for a specific sample and sample set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_samples_bar.html#malariagen_data.ag3.Ag3.plot_samples_bar" title="malariagen_data.ag3.Ag3.plot_samples_bar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_samples_bar</span></code></a>(x[, color, sort, ...])</p></td>
<td><p>Plot a bar chart showing the number of samples available, grouped by some variable such as country or year.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_samples_interactive_map.html#malariagen_data.ag3.Ag3.plot_samples_interactive_map" title="malariagen_data.ag3.Ag3.plot_samples_interactive_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_samples_interactive_map</span></code></a>([sample_sets, ...])</p></td>
<td><p>Plot an interactive map showing sampling locations using ipyleaflet.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_sample_location_mapbox.html#malariagen_data.ag3.Ag3.plot_sample_location_mapbox" title="malariagen_data.ag3.Ag3.plot_sample_location_mapbox"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_sample_location_mapbox</span></code></a>(*, sample_sets)</p></td>
<td><p>Plot markers on a map showing sample locations as a Mapbox scatter plot.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_sample_location_geo.html#malariagen_data.ag3.Ag3.plot_sample_location_geo" title="malariagen_data.ag3.Ag3.plot_sample_location_geo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_sample_location_geo</span></code></a>(*, sample_sets[, ...])</p></td>
<td><p>Plot markers on a map showing sample locations as a geographic scatter plot.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.wgs_data_catalog.html#malariagen_data.ag3.Ag3.wgs_data_catalog" title="malariagen_data.ag3.Ag3.wgs_data_catalog"><code class="xref py py-obj docutils literal notranslate"><span class="pre">wgs_data_catalog</span></code></a>(sample_set)</p></td>
<td><p>Load a data catalog providing URLs for downloading BAM, VCF and Zarr files for samples in a given sample set.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.cohorts.html#malariagen_data.ag3.Ag3.cohorts" title="malariagen_data.ag3.Ag3.cohorts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cohorts</span></code></a>(cohort_set)</p></td>
<td><p>Read data for a specific cohort set, including cohort size, country code, taxon, administrative units name, ISO code, geoBoundaries shape ID and representative latitude and longitude points.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="snp-data-access">
<h2>SNP data access<a class="headerlink" href="#snp-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.site_mask_ids.html#malariagen_data.ag3.Ag3.site_mask_ids" title="malariagen_data.ag3.Ag3.site_mask_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">site_mask_ids</span></code></a></p></td>
<td><p>Identifiers for the different site masks that are available.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.snp_calls.html#malariagen_data.ag3.Ag3.snp_calls" title="malariagen_data.ag3.Ag3.snp_calls"><code class="xref py py-obj docutils literal notranslate"><span class="pre">snp_calls</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Access SNP sites, site filters and genotype calls.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.snp_allele_counts.html#malariagen_data.ag3.Ag3.snp_allele_counts" title="malariagen_data.ag3.Ag3.snp_allele_counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">snp_allele_counts</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Compute SNP allele counts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_snps.html#malariagen_data.ag3.Ag3.plot_snps" title="malariagen_data.ag3.Ag3.plot_snps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_snps</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Plot SNPs in a given genome region.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.site_annotations.html#malariagen_data.ag3.Ag3.site_annotations" title="malariagen_data.ag3.Ag3.site_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">site_annotations</span></code></a>(region[, site_mask, ...])</p></td>
<td><p>Load site annotations.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.is_accessible.html#malariagen_data.ag3.Ag3.is_accessible" title="malariagen_data.ag3.Ag3.is_accessible"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_accessible</span></code></a>(region[, site_mask, ...])</p></td>
<td><p>Compute genome accessibility array.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.biallelic_snp_calls.html#malariagen_data.ag3.Ag3.biallelic_snp_calls" title="malariagen_data.ag3.Ag3.biallelic_snp_calls"><code class="xref py py-obj docutils literal notranslate"><span class="pre">biallelic_snp_calls</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Access SNP calls at sites which are biallelic within the selected samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.biallelic_diplotypes.html#malariagen_data.ag3.Ag3.biallelic_diplotypes" title="malariagen_data.ag3.Ag3.biallelic_diplotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">biallelic_diplotypes</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Load biallelic SNP genotypes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.biallelic_snps_to_plink.html#malariagen_data.ag3.Ag3.biallelic_snps_to_plink" title="malariagen_data.ag3.Ag3.biallelic_snps_to_plink"><code class="xref py py-obj docutils literal notranslate"><span class="pre">biallelic_snps_to_plink</span></code></a>(output_dir, region, ...)</p></td>
<td><p>Write Anopheles biallelic SNP data to the Plink binary file format.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="haplotype-data-access">
<h2>Haplotype data access<a class="headerlink" href="#haplotype-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.phasing_analysis_ids.html#malariagen_data.ag3.Ag3.phasing_analysis_ids" title="malariagen_data.ag3.Ag3.phasing_analysis_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phasing_analysis_ids</span></code></a></p></td>
<td><p>Identifiers for the different phasing analyses that are available.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.haplotypes.html#malariagen_data.ag3.Ag3.haplotypes" title="malariagen_data.ag3.Ag3.haplotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">haplotypes</span></code></a>(region[, analysis, sample_sets, ...])</p></td>
<td><p>Access haplotype data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.haplotype_sites.html#malariagen_data.ag3.Ag3.haplotype_sites" title="malariagen_data.ag3.Ag3.haplotype_sites"><code class="xref py py-obj docutils literal notranslate"><span class="pre">haplotype_sites</span></code></a>(region, field[, analysis, ...])</p></td>
<td><p>Access haplotype site data (positions or alleles).</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="aim-data-access">
<h2>AIM data access<a class="headerlink" href="#aim-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.aim_ids.html#malariagen_data.ag3.Ag3.aim_ids" title="malariagen_data.ag3.Ag3.aim_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aim_ids</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.aim_variants.html#malariagen_data.ag3.Ag3.aim_variants" title="malariagen_data.ag3.Ag3.aim_variants"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aim_variants</span></code></a>(aims)</p></td>
<td><p>Access ancestry informative marker variants.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.aim_calls.html#malariagen_data.ag3.Ag3.aim_calls" title="malariagen_data.ag3.Ag3.aim_calls"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aim_calls</span></code></a>(aims[, sample_sets, sample_query, ...])</p></td>
<td><p>Access ancestry informative marker SNP sites, alleles and genotype calls.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_aim_heatmap.html#malariagen_data.ag3.Ag3.plot_aim_heatmap" title="malariagen_data.ag3.Ag3.plot_aim_heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_aim_heatmap</span></code></a>(aims[, sample_sets, ...])</p></td>
<td><p>Plot a heatmap of ancestry-informative marker (AIM) genotypes.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="cnv-data-access">
<h2>CNV data access<a class="headerlink" href="#cnv-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.coverage_calls_analysis_ids.html#malariagen_data.ag3.Ag3.coverage_calls_analysis_ids" title="malariagen_data.ag3.Ag3.coverage_calls_analysis_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">coverage_calls_analysis_ids</span></code></a></p></td>
<td><p>Identifiers for the different coverage calls analyses that are available.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.cnv_hmm.html#malariagen_data.ag3.Ag3.cnv_hmm" title="malariagen_data.ag3.Ag3.cnv_hmm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cnv_hmm</span></code></a>(region[, sample_sets, sample_query, ...])</p></td>
<td><p>Access CNV HMM data from CNV calling.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.cnv_coverage_calls.html#malariagen_data.ag3.Ag3.cnv_coverage_calls" title="malariagen_data.ag3.Ag3.cnv_coverage_calls"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cnv_coverage_calls</span></code></a>(region, sample_set, analysis)</p></td>
<td><p>Access CNV HMM data from genome-wide CNV discovery and filtering.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.cnv_discordant_read_calls.html#malariagen_data.ag3.Ag3.cnv_discordant_read_calls" title="malariagen_data.ag3.Ag3.cnv_discordant_read_calls"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cnv_discordant_read_calls</span></code></a>(contig[, ...])</p></td>
<td><p>Access CNV discordant read calls data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_cnv_hmm_coverage.html#malariagen_data.ag3.Ag3.plot_cnv_hmm_coverage" title="malariagen_data.ag3.Ag3.plot_cnv_hmm_coverage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_cnv_hmm_coverage</span></code></a>(sample, region[, ...])</p></td>
<td><p>Plot CNV HMM data for a single sample, together with a genes track, using bokeh.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_cnv_hmm_heatmap.html#malariagen_data.ag3.Ag3.plot_cnv_hmm_heatmap" title="malariagen_data.ag3.Ag3.plot_cnv_hmm_heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_cnv_hmm_heatmap</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Plot CNV HMM data for multiple samples as a heatmap, with a genes track, using bokeh.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.gene_cnv.html#malariagen_data.ag3.Ag3.gene_cnv" title="malariagen_data.ag3.Ag3.gene_cnv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gene_cnv</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Compute modal copy number by gene, from HMM data.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="integrative-genomics-viewer-igv">
<h2>Integrative genomics viewer (IGV)<a class="headerlink" href="#integrative-genomics-viewer-igv" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.igv.html#malariagen_data.ag3.Ag3.igv" title="malariagen_data.ag3.Ag3.igv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">igv</span></code></a>(region[, tracks, init])</p></td>
<td><p>Create an IGV browser and inject into the current notebook.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.view_alignments.html#malariagen_data.ag3.Ag3.view_alignments" title="malariagen_data.ag3.Ag3.view_alignments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">view_alignments</span></code></a>(region, sample[, ...])</p></td>
<td><p>Launch IGV and view sequence read alignments and SNP genotypes from the given sample.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="snp-and-cnv-frequency-analysis">
<h2>SNP and CNV frequency analysis<a class="headerlink" href="#snp-and-cnv-frequency-analysis" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.snp_allele_frequencies.html#malariagen_data.ag3.Ag3.snp_allele_frequencies" title="malariagen_data.ag3.Ag3.snp_allele_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">snp_allele_frequencies</span></code></a>(transcript, cohorts)</p></td>
<td><p>Compute SNP allele frequencies for a gene transcript.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.snp_allele_frequencies_advanced.html#malariagen_data.ag3.Ag3.snp_allele_frequencies_advanced" title="malariagen_data.ag3.Ag3.snp_allele_frequencies_advanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">snp_allele_frequencies_advanced</span></code></a>(transcript, ...)</p></td>
<td><p>Group samples by taxon, area (space) and period (time), then compute SNP allele frequencies.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.aa_allele_frequencies.html#malariagen_data.ag3.Ag3.aa_allele_frequencies" title="malariagen_data.ag3.Ag3.aa_allele_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aa_allele_frequencies</span></code></a>(transcript, cohorts[, ...])</p></td>
<td><p>Compute amino acid substitution frequencies for a gene transcript.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.aa_allele_frequencies_advanced.html#malariagen_data.ag3.Ag3.aa_allele_frequencies_advanced" title="malariagen_data.ag3.Ag3.aa_allele_frequencies_advanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aa_allele_frequencies_advanced</span></code></a>(transcript, ...)</p></td>
<td><p>Group samples by taxon, area (space) and period (time), then compute amino acid change allele frequencies.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.gene_cnv_frequencies.html#malariagen_data.ag3.Ag3.gene_cnv_frequencies" title="malariagen_data.ag3.Ag3.gene_cnv_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gene_cnv_frequencies</span></code></a>(region, cohorts[, ...])</p></td>
<td><p>Compute modal copy number by gene, then compute the frequency of amplifications and deletions in one or more cohorts, from HMM data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.gene_cnv_frequencies_advanced.html#malariagen_data.ag3.Ag3.gene_cnv_frequencies_advanced" title="malariagen_data.ag3.Ag3.gene_cnv_frequencies_advanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gene_cnv_frequencies_advanced</span></code></a>(region, ...[, ...])</p></td>
<td><p>Group samples by taxon, area (space) and period (time), then compute gene CNV counts and frequencies.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.haplotypes_frequencies.html#malariagen_data.ag3.Ag3.haplotypes_frequencies" title="malariagen_data.ag3.Ag3.haplotypes_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">haplotypes_frequencies</span></code></a>(region, cohorts[, ...])</p></td>
<td><p>Compute haplotype frequencies for a region.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.haplotypes_frequencies_advanced.html#malariagen_data.ag3.Ag3.haplotypes_frequencies_advanced" title="malariagen_data.ag3.Ag3.haplotypes_frequencies_advanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">haplotypes_frequencies_advanced</span></code></a>(region, ...)</p></td>
<td><p>Group samples by taxon, area (space) and period (time), then compute haplotype frequencies.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_frequencies_heatmap.html#malariagen_data.ag3.Ag3.plot_frequencies_heatmap" title="malariagen_data.ag3.Ag3.plot_frequencies_heatmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_frequencies_heatmap</span></code></a>(df[, index, ...])</p></td>
<td><p>Plot a heatmap from a pandas DataFrame of frequencies, e.g., output from <cite>snp_allele_frequencies()</cite> or <cite>gene_cnv_frequencies()</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_frequencies_time_series.html#malariagen_data.ag3.Ag3.plot_frequencies_time_series" title="malariagen_data.ag3.Ag3.plot_frequencies_time_series"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_frequencies_time_series</span></code></a>(ds[, height, ...])</p></td>
<td><p>Create a time series plot of variant frequencies using plotly.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_frequencies_interactive_map.html#malariagen_data.ag3.Ag3.plot_frequencies_interactive_map" title="malariagen_data.ag3.Ag3.plot_frequencies_interactive_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_frequencies_interactive_map</span></code></a>(ds[, ...])</p></td>
<td><p>Create an interactive map with markers showing variant frequencies or cohorts grouped by area (space), period (time) and taxon.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="principal-components-analysis-pca">
<h2>Principal components analysis (PCA)<a class="headerlink" href="#principal-components-analysis-pca" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.pca.html#malariagen_data.ag3.Ag3.pca" title="malariagen_data.ag3.Ag3.pca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pca</span></code></a>(region, n_snps[, n_components, ...])</p></td>
<td><p>Run a principal components analysis (PCA) using biallelic SNPs from the selected genome region and samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_pca_variance.html#malariagen_data.ag3.Ag3.plot_pca_variance" title="malariagen_data.ag3.Ag3.plot_pca_variance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_pca_variance</span></code></a>(evr[, width, height, ...])</p></td>
<td><p>Plot explained variance ratios from a principal components analysis (PCA) using a plotly bar plot.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_pca_coords.html#malariagen_data.ag3.Ag3.plot_pca_coords" title="malariagen_data.ag3.Ag3.plot_pca_coords"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_pca_coords</span></code></a>(data[, x, y, color, symbol, ...])</p></td>
<td><p>Plot sample coordinates from a principal components analysis (PCA) as a plotly scatter plot.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_pca_coords_3d.html#malariagen_data.ag3.Ag3.plot_pca_coords_3d" title="malariagen_data.ag3.Ag3.plot_pca_coords_3d"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_pca_coords_3d</span></code></a>(data[, x, y, z, color, ...])</p></td>
<td><p>Plot sample coordinates from a principal components analysis (PCA) as a plotly 3D scatter plot.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="genetic-distance-and-neighbour-joining-trees-njt">
<h2>Genetic distance and neighbour-joining trees (NJT)<a class="headerlink" href="#genetic-distance-and-neighbour-joining-trees-njt" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_njt.html#malariagen_data.ag3.Ag3.plot_njt" title="malariagen_data.ag3.Ag3.plot_njt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_njt</span></code></a>(region, n_snps[, color, symbol, ...])</p></td>
<td><p>Plot an unrooted neighbour-joining tree, computed from pairwise distances between samples using biallelic SNP genotypes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.njt.html#malariagen_data.ag3.Ag3.njt" title="malariagen_data.ag3.Ag3.njt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">njt</span></code></a>(region, n_snps[, algorithm, metric, ...])</p></td>
<td><p>Construct a neighbour-joining tree between samples using biallelic SNP genotypes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.biallelic_diplotype_pairwise_distances.html#malariagen_data.ag3.Ag3.biallelic_diplotype_pairwise_distances" title="malariagen_data.ag3.Ag3.biallelic_diplotype_pairwise_distances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">biallelic_diplotype_pairwise_distances</span></code></a>(...)</p></td>
<td><p>Compute pairwise distances between samples using biallelic SNP genotypes.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="heterozygosity-analysis">
<h2>Heterozygosity analysis<a class="headerlink" href="#heterozygosity-analysis" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_heterozygosity.html#malariagen_data.ag3.Ag3.plot_heterozygosity" title="malariagen_data.ag3.Ag3.plot_heterozygosity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_heterozygosity</span></code></a>(sample, region[, ...])</p></td>
<td><p>Plot windowed heterozygosity for a single sample over a genome region.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.roh_hmm.html#malariagen_data.ag3.Ag3.roh_hmm" title="malariagen_data.ag3.Ag3.roh_hmm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">roh_hmm</span></code></a>(sample, region[, window_size, ...])</p></td>
<td><p>Infer runs of homozygosity for a single sample over a genome region.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_roh.html#malariagen_data.ag3.Ag3.plot_roh" title="malariagen_data.ag3.Ag3.plot_roh"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_roh</span></code></a>(sample, region[, window_size, ...])</p></td>
<td><p>Plot windowed heterozygosity and inferred runs of homozygosity for a single sample over a genome region.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="diversity-analysis">
<h2>Diversity analysis<a class="headerlink" href="#diversity-analysis" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.cohort_diversity_stats.html#malariagen_data.ag3.Ag3.cohort_diversity_stats" title="malariagen_data.ag3.Ag3.cohort_diversity_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cohort_diversity_stats</span></code></a>(cohort, cohort_size, ...)</p></td>
<td><p>Compute genetic diversity summary statistics for a cohort of individuals.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.diversity_stats.html#malariagen_data.ag3.Ag3.diversity_stats" title="malariagen_data.ag3.Ag3.diversity_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diversity_stats</span></code></a>(cohorts, cohort_size, region)</p></td>
<td><p>Compute genetic diversity summary statistics for multiple cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_diversity_stats.html#malariagen_data.ag3.Ag3.plot_diversity_stats" title="malariagen_data.ag3.Ag3.plot_diversity_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_diversity_stats</span></code></a>(df_stats[, color, ...])</p></td>
<td><p>Plot diversity summary statistics for multiple cohorts.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="genome-wide-selection-scans">
<h2>Genome-wide selection scans<a class="headerlink" href="#genome-wide-selection-scans" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.h12_calibration.html#malariagen_data.ag3.Ag3.h12_calibration" title="malariagen_data.ag3.Ag3.h12_calibration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">h12_calibration</span></code></a>(contig[, analysis, ...])</p></td>
<td><p>Generate h12 GWSS calibration data for different window sizes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_h12_calibration.html#malariagen_data.ag3.Ag3.plot_h12_calibration" title="malariagen_data.ag3.Ag3.plot_h12_calibration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_h12_calibration</span></code></a>(contig[, analysis, ...])</p></td>
<td><p>Plot h12 GWSS calibration data for different window sizes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.h12_gwss.html#malariagen_data.ag3.Ag3.h12_gwss" title="malariagen_data.ag3.Ag3.h12_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">h12_gwss</span></code></a>(contig, window_size[, analysis, ...])</p></td>
<td><p>Run h12 genome-wide selection scan.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_h12_gwss.html#malariagen_data.ag3.Ag3.plot_h12_gwss" title="malariagen_data.ag3.Ag3.plot_h12_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_h12_gwss</span></code></a>(contig, window_size[, ...])</p></td>
<td><p>Plot h12 GWSS data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_h12_gwss_multi_panel.html#malariagen_data.ag3.Ag3.plot_h12_gwss_multi_panel" title="malariagen_data.ag3.Ag3.plot_h12_gwss_multi_panel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_h12_gwss_multi_panel</span></code></a>(contig, cohorts, ...)</p></td>
<td><p>Plot h12 GWSS data with multiple tracks.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_h12_gwss_multi_overlay.html#malariagen_data.ag3.Ag3.plot_h12_gwss_multi_overlay" title="malariagen_data.ag3.Ag3.plot_h12_gwss_multi_overlay"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_h12_gwss_multi_overlay</span></code></a>(contig, cohorts, ...)</p></td>
<td><p>Plot h12 GWSS data with multiple traces overlaid.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.h1x_gwss.html#malariagen_data.ag3.Ag3.h1x_gwss" title="malariagen_data.ag3.Ag3.h1x_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">h1x_gwss</span></code></a>(contig, window_size, cohort1_query, ...)</p></td>
<td><p>Run a H1X genome-wide scan to detect genome regions with shared selective sweeps between two cohorts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_h1x_gwss.html#malariagen_data.ag3.Ag3.plot_h1x_gwss" title="malariagen_data.ag3.Ag3.plot_h1x_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_h1x_gwss</span></code></a>(contig, window_size, ...[, ...])</p></td>
<td><p>Run and plot a H1X genome-wide scan to detect genome regions with shared selective sweeps between two cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.g123_calibration.html#malariagen_data.ag3.Ag3.g123_calibration" title="malariagen_data.ag3.Ag3.g123_calibration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">g123_calibration</span></code></a>(contig[, sites, site_mask, ...])</p></td>
<td><p>Generate G123 GWSS calibration data for different window sizes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_g123_calibration.html#malariagen_data.ag3.Ag3.plot_g123_calibration" title="malariagen_data.ag3.Ag3.plot_g123_calibration"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_g123_calibration</span></code></a>(contig, sites[, ...])</p></td>
<td><p>Plot G123 GWSS calibration data for different window sizes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.g123_gwss.html#malariagen_data.ag3.Ag3.g123_gwss" title="malariagen_data.ag3.Ag3.g123_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">g123_gwss</span></code></a>(contig, window_size[, sites, ...])</p></td>
<td><p>Run a G123 genome-wide selection scan.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_g123_gwss.html#malariagen_data.ag3.Ag3.plot_g123_gwss" title="malariagen_data.ag3.Ag3.plot_g123_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_g123_gwss</span></code></a>(contig, window_size[, sites, ...])</p></td>
<td><p>Plot G123 GWSS data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.ihs_gwss.html#malariagen_data.ag3.Ag3.ihs_gwss" title="malariagen_data.ag3.Ag3.ihs_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ihs_gwss</span></code></a>(contig[, analysis, sample_sets, ...])</p></td>
<td><p>Run iHS GWSS.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_ihs_gwss.html#malariagen_data.ag3.Ag3.plot_ihs_gwss" title="malariagen_data.ag3.Ag3.plot_ihs_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_ihs_gwss</span></code></a>(contig[, analysis, ...])</p></td>
<td><p>Run and plot iHS GWSS data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.xpehh_gwss.html#malariagen_data.ag3.Ag3.xpehh_gwss" title="malariagen_data.ag3.Ag3.xpehh_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">xpehh_gwss</span></code></a>(contig[, analysis, sample_sets, ...])</p></td>
<td><p>Run XP-EHH GWSS.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_xpehh_gwss.html#malariagen_data.ag3.Ag3.plot_xpehh_gwss" title="malariagen_data.ag3.Ag3.plot_xpehh_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_xpehh_gwss</span></code></a>(contig[, analysis, ...])</p></td>
<td><p>Run and plot XP-EHH GWSS data.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="haplotype-clustering-and-network-analysis">
<h2>Haplotype clustering and network analysis<a class="headerlink" href="#haplotype-clustering-and-network-analysis" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_haplotype_clustering.html#malariagen_data.ag3.Ag3.plot_haplotype_clustering" title="malariagen_data.ag3.Ag3.plot_haplotype_clustering"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_haplotype_clustering</span></code></a>(region[, ...])</p></td>
<td><p>Hierarchically cluster haplotypes in region and produce an interactive plot.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_haplotype_network.html#malariagen_data.ag3.Ag3.plot_haplotype_network" title="malariagen_data.ag3.Ag3.plot_haplotype_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_haplotype_network</span></code></a>(region[, analysis, ...])</p></td>
<td><p>Construct a median-joining haplotype network and display it using Cytoscape.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.haplotype_pairwise_distances.html#malariagen_data.ag3.Ag3.haplotype_pairwise_distances" title="malariagen_data.ag3.Ag3.haplotype_pairwise_distances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">haplotype_pairwise_distances</span></code></a>(region[, ...])</p></td>
<td><p>Compute pairwise distances between haplotypes.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="diplotype-clustering">
<h2>Diplotype clustering<a class="headerlink" href="#diplotype-clustering" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_diplotype_clustering.html#malariagen_data.ag3.Ag3.plot_diplotype_clustering" title="malariagen_data.ag3.Ag3.plot_diplotype_clustering"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_diplotype_clustering</span></code></a>(region[, ...])</p></td>
<td><p>&quot;         Hierarchically cluster diplotypes in region and produce an interactive plot.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_diplotype_clustering_advanced.html#malariagen_data.ag3.Ag3.plot_diplotype_clustering_advanced" title="malariagen_data.ag3.Ag3.plot_diplotype_clustering_advanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_diplotype_clustering_advanced</span></code></a>(region[, ...])</p></td>
<td><p>&quot;         Perform diplotype clustering, annotated with heterozygosity, gene copy number and amino acid variants.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="fst-analysis">
<h2>Fst analysis<a class="headerlink" href="#fst-analysis" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.average_fst.html#malariagen_data.ag3.Ag3.average_fst" title="malariagen_data.ag3.Ag3.average_fst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">average_fst</span></code></a>(region, cohort1_query, cohort2_query)</p></td>
<td><p>Compute average Hudson's Fst between two specified cohorts.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.pairwise_average_fst.html#malariagen_data.ag3.Ag3.pairwise_average_fst" title="malariagen_data.ag3.Ag3.pairwise_average_fst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pairwise_average_fst</span></code></a>(region, cohorts[, ...])</p></td>
<td><p>Compute pairwise average Hudson's Fst between a set of specified cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_pairwise_average_fst.html#malariagen_data.ag3.Ag3.plot_pairwise_average_fst" title="malariagen_data.ag3.Ag3.plot_pairwise_average_fst"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_pairwise_average_fst</span></code></a>(fst_df[, ...])</p></td>
<td><p>Plot a heatmap of pairwise average Fst values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.fst_gwss.html#malariagen_data.ag3.Ag3.fst_gwss" title="malariagen_data.ag3.Ag3.fst_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fst_gwss</span></code></a>(contig, window_size, cohort1_query, ...)</p></td>
<td><p>Run a Fst genome-wide scan to investigate genetic differentiation between two cohorts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.plot_fst_gwss.html#malariagen_data.ag3.Ag3.plot_fst_gwss" title="malariagen_data.ag3.Ag3.plot_fst_gwss"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_fst_gwss</span></code></a>(contig, window_size, ...[, ...])</p></td>
<td><p>Run and plot a Fst genome-wide scan to investigate genetic differentiation between two cohorts.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="inversion-karyotypes">
<h2>Inversion karyotypes<a class="headerlink" href="#inversion-karyotypes" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.karyotype.html#malariagen_data.ag3.Ag3.karyotype" title="malariagen_data.ag3.Ag3.karyotype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">karyotype</span></code></a>(inversion[, sample_sets, ...])</p></td>
<td><p>Infer karyotype from tag SNPs for a given inversion in Ag.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="phenotype-data-access">
<h2>Phenotype data access<a class="headerlink" href="#phenotype-data-access" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.phenotype_data.html#malariagen_data.ag3.Ag3.phenotype_data" title="malariagen_data.ag3.Ag3.phenotype_data"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phenotype_data</span></code></a>([sample_sets, sample_query, ...])</p></td>
<td><p>Load phenotypic data from insecticide resistance bioassays.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.phenotypes_with_snps.html#malariagen_data.ag3.Ag3.phenotypes_with_snps" title="malariagen_data.ag3.Ag3.phenotypes_with_snps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phenotypes_with_snps</span></code></a>(region[, sample_sets, ...])</p></td>
<td><p>Combine phenotypic traits with SNP genotype data for GWAS analysis.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.phenotypes_with_haplotypes.html#malariagen_data.ag3.Ag3.phenotypes_with_haplotypes" title="malariagen_data.ag3.Ag3.phenotypes_with_haplotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phenotypes_with_haplotypes</span></code></a>(region[, ...])</p></td>
<td><p>Combine phenotypic traits with haplotype data for extended association analysis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.phenotype_sample_sets.html#malariagen_data.ag3.Ag3.phenotype_sample_sets" title="malariagen_data.ag3.Ag3.phenotype_sample_sets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phenotype_sample_sets</span></code></a>()</p></td>
<td><p>List sample sets that contain phenotypic data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/malariagen_data.ag3.Ag3.phenotype_binary.html#malariagen_data.ag3.Ag3.phenotype_binary" title="malariagen_data.ag3.Ag3.phenotype_binary"><code class="xref py py-obj docutils literal notranslate"><span class="pre">phenotype_binary</span></code></a>([sample_sets, insecticide, ...])</p></td>
<td><p>Convert phenotype data into binary format for statistical analysis.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-data-access">Basic data access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#reference-genome-data-access">Reference genome data access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-metadata-access">Sample metadata access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#snp-data-access">SNP data access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#haplotype-data-access">Haplotype data access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#aim-data-access">AIM data access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cnv-data-access">CNV data access</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#integrative-genomics-viewer-igv">Integrative genomics viewer (IGV)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#snp-and-cnv-frequency-analysis">SNP and CNV frequency analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#principal-components-analysis-pca">Principal components analysis (PCA)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-distance-and-neighbour-joining-trees-njt">Genetic distance and neighbour-joining trees (NJT)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#heterozygosity-analysis">Heterozygosity analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#diversity-analysis">Diversity analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genome-wide-selection-scans">Genome-wide selection scans</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#haplotype-clustering-and-network-analysis">Haplotype clustering and network analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#diplotype-clustering">Diplotype clustering</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#fst-analysis">Fst analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#inversion-karyotypes">Inversion karyotypes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#phenotype-data-access">Phenotype data access</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Ag3.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2024, MalariaGEN.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>