{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cef10418-08a1-4cc4-a3b0-b8b516cc64fd",
   "metadata": {},
   "source": [
    "# Demonstrating New Phenotype Data Access and Filtering\n",
    "\n",
    "This notebook showcases the updated functionalities for accessing and filtering insecticide resistance phenotype data, and merging it with genetic data (SNPs and Haplotypes), as per the recent refactoring.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29554ea2-5e37-40b6-90c1-08088bf9793a",
   "metadata": {},
   "source": [
    "## 1. Setup and Imports\n",
    "\n",
    "First, we'll import the necessary libraries and instantiate the `Ag3` API client."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f1280de7-129f-4875-bf12-eb68b8d88573",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "'use strict';\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    function drop(id) {\n",
       "      const view = Bokeh.index.get_by_id(id)\n",
       "      if (view != null) {\n",
       "        view.model.document.clear()\n",
       "        Bokeh.index.delete(view)\n",
       "      }\n",
       "    }\n",
       "\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null) {\n",
       "      drop(id)\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim()\n",
       "            drop(id)\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded(error = null) {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      const html = (() => {\n",
       "        if (typeof root.Bokeh === \"undefined\") {\n",
       "          if (error == null) {\n",
       "            return \"BokehJS is loading ...\";\n",
       "          } else {\n",
       "            return \"BokehJS failed to load.\";\n",
       "          }\n",
       "        } else {\n",
       "          const prefix = `BokehJS ${root.Bokeh.version}`;\n",
       "          if (error == null) {\n",
       "            return `${prefix} successfully loaded.`;\n",
       "          } else {\n",
       "            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n",
       "          }\n",
       "        }\n",
       "      })();\n",
       "      el.innerHTML = html;\n",
       "\n",
       "      if (error != null) {\n",
       "        const wrapper = document.createElement(\"div\");\n",
       "        wrapper.style.overflow = \"auto\";\n",
       "        wrapper.style.height = \"5em\";\n",
       "        wrapper.style.resize = \"vertical\";\n",
       "        const content = document.createElement(\"div\");\n",
       "        content.style.fontFamily = \"monospace\";\n",
       "        content.style.whiteSpace = \"pre-wrap\";\n",
       "        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n",
       "        content.textContent = error.stack ?? error.toString();\n",
       "        wrapper.append(content);\n",
       "        el.append(wrapper);\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(() => display_loaded(error), 100);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.6.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      try {\n",
       "            for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "\n",
       "      } catch (error) {throw error;\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "'use strict';\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded(error = null) {\n    const el = document.getElementById(null);\n    if (el != null) {\n      const html = (() => {\n        if (typeof root.Bokeh === \"undefined\") {\n          if (error == null) {\n            return \"BokehJS is loading ...\";\n          } else {\n            return \"BokehJS failed to load.\";\n          }\n        } else {\n          const prefix = `BokehJS ${root.Bokeh.version}`;\n          if (error == null) {\n            return `${prefix} successfully loaded.`;\n          } else {\n            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n          }\n        }\n      })();\n      el.innerHTML = html;\n\n      if (error != null) {\n        const wrapper = document.createElement(\"div\");\n        wrapper.style.overflow = \"auto\";\n        wrapper.style.height = \"5em\";\n        wrapper.style.resize = \"vertical\";\n        const content = document.createElement(\"div\");\n        content.style.fontFamily = \"monospace\";\n        content.style.whiteSpace = \"pre-wrap\";\n        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n        content.textContent = error.stack ?? error.toString();\n        wrapper.append(content);\n        el.append(wrapper);\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(() => display_loaded(error), 100);\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.6.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.6.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n      try {\n            for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n\n      } catch (error) {throw error;\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MalariaGEN Ag3 API client initialized.\n",
      "<MalariaGEN Ag3 API client>\n",
      "Storage URL             : gs://vo_agam_release_master_us_central1/\n",
      "Data releases available : 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 3.10, 3.11, 3.12, 3.13, 3.14, 3.15\n",
      "Results cache           : None\n",
      "Cohorts analysis        : 20250502\n",
      "AIM analysis            : 20220528\n",
      "Site filters analysis   : dt_20200416\n",
      "Software version        : malariagen_data 15.2.2.post22+48d1097a\n",
      "Client location         : Tanger-Tetouan-Al Hoceima, Morocco\n",
      "---\n",
      "Please note that data are subject to terms of use,\n",
      "for more information see https://www.malariagen.net/data\n",
      "or contact support@malariagen.net. For API documentation see \n",
      "https://malariagen.github.io/malariagen-data-python/v15.2.2.post22+48d1097a/Ag3.html\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import xarray as xr\n",
    "import warnings\n",
    "from malariagen_data import Ag3\n",
    "\n",
    "# Suppress warnings for cleaner output in the notebook\n",
    "warnings.simplefilter(\"ignore\", UserWarning)\n",
    "\n",
    "# Instantiate the Ag3 API client\n",
    "# Use pre=True for development versions, or specify a release like release=\"v3.2\"\n",
    "ag3 = Ag3(pre=True)\n",
    "\n",
    "print(\"MalariaGEN Ag3 API client initialized.\")\n",
    "print(ag3)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b1975f3-ad03-450d-b387-6d2ffd5c0143",
   "metadata": {},
   "source": [
    "## 2. Listing Available Phenotype Sample Sets\n",
    "Let's see which sample sets have phenotype data available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7dd3ebc4-64f7-4d0b-ad01-fbc1e08b0636",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available phenotype sample sets: ['1237-VO-BJ-DJOGBENOU-VMF00050', '1244-VO-GH-YAWSON-VMF00051', '1245-VO-CI-CONSTANT-VMF00054', '1253-VO-TG-DJOGBENOU-VMF00052']\n",
      "\n",
      "Using sample set for demonstration: 1237-VO-BJ-DJOGBENOU-VMF00050\n"
     ]
    }
   ],
   "source": [
    "phenotype_sample_sets = ag3.phenotype_sample_sets()\n",
    "print(f\"Available phenotype sample sets: {phenotype_sample_sets}\")\n",
    "\n",
    "# We'll pick one sample set for demonstration, preferably one known to have data\n",
    "# For this example, we'll use '1237-VO-BJ-DJOGBENOU-VMF00050'\n",
    "demo_sample_set = '1237-VO-BJ-DJOGBENOU-VMF00050'\n",
    "if demo_sample_set not in phenotype_sample_sets:\n",
    "    print(f\"Warning: '{demo_sample_set}' not found. Using the first available: {phenotype_sample_sets}\")\n",
    "    demo_sample_set = phenotype_sample_sets\n",
    "\n",
    "print(f\"\\nUsing sample set for demonstration: {demo_sample_set}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1319b97a-bd6f-4246-b653-a394aee97930",
   "metadata": {},
   "source": [
    "## 3. Loading Phenotype Data with sample_query\n",
    "The phenotype_data() method now exclusively returns a Pandas DataFrame containing phenotype and sample metadata. Filtering by attributes like insecticide, location, dose, etc., is done using the sample_query parameter, which accepts a Pandas-style query string."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6d5023-17b1-4988-8683-22b3f71096ff",
   "metadata": {},
   "source": [
    "### Example 1: Filter by a specific insecticide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0b9d0ae-6a79-458a-ad9f-9281a1c6112e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Loading phenotype data for '1237-VO-BJ-DJOGBENOU-VMF00050' filtered by Deltamethrin ---\n",
      "Shape of DataFrame: (88, 60)         \n",
      "\n",
      "First 5 rows of the filtered DataFrame:\n",
      "\n",
      "Unique insecticides in filtered data: ['Deltamethrin']\n",
      "\n",
      "DataFrame Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 88 entries, 0 to 87\n",
      "Data columns (total 60 columns):\n",
      " #   Column                           Non-Null Count  Dtype  \n",
      "---  ------                           --------------  -----  \n",
      " 0   sample_id                        88 non-null     object \n",
      " 1   country                          88 non-null     object \n",
      " 2   location                         88 non-null     object \n",
      " 3   insecticide                      88 non-null     object \n",
      " 4   dose                             88 non-null     float64\n",
      " 5   phenotype                        88 non-null     object \n",
      " 6   sample_set                       88 non-null     object \n",
      " 7   partner_sample_id                88 non-null     object \n",
      " 8   contributor                      88 non-null     object \n",
      " 9   year                             88 non-null     int64  \n",
      " 10  month                            88 non-null     int64  \n",
      " 11  latitude                         88 non-null     float64\n",
      " 12  longitude                        88 non-null     float64\n",
      " 13  sex_call                         88 non-null     object \n",
      " 14  release                          88 non-null     object \n",
      " 15  quarter                          88 non-null     int64  \n",
      " 16  study_id                         88 non-null     object \n",
      " 17  study_url                        88 non-null     object \n",
      " 18  terms_of_use_expiry_date         88 non-null     object \n",
      " 19  terms_of_use_url                 88 non-null     object \n",
      " 20  unrestricted_use                 88 non-null     bool   \n",
      " 21  mean_cov                         88 non-null     Float64\n",
      " 22  median_cov                       88 non-null     Int64  \n",
      " 23  modal_cov                        88 non-null     Int64  \n",
      " 24  mean_cov_2L                      88 non-null     Float64\n",
      " 25  median_cov_2L                    88 non-null     Int64  \n",
      " 26  mode_cov_2L                      88 non-null     Int64  \n",
      " 27  mean_cov_2R                      88 non-null     Float64\n",
      " 28  median_cov_2R                    88 non-null     Int64  \n",
      " 29  mode_cov_2R                      88 non-null     Int64  \n",
      " 30  mean_cov_3L                      88 non-null     Float64\n",
      " 31  median_cov_3L                    88 non-null     Int64  \n",
      " 32  mode_cov_3L                      88 non-null     Int64  \n",
      " 33  mean_cov_3R                      88 non-null     Float64\n",
      " 34  median_cov_3R                    88 non-null     Int64  \n",
      " 35  mode_cov_3R                      88 non-null     Int64  \n",
      " 36  mean_cov_X                       88 non-null     Float64\n",
      " 37  median_cov_X                     88 non-null     Int64  \n",
      " 38  mode_cov_X                       88 non-null     Int64  \n",
      " 39  frac_gen_cov                     88 non-null     Float64\n",
      " 40  divergence                       88 non-null     Float64\n",
      " 41  contam_pct                       88 non-null     Float64\n",
      " 42  contam_LLR                       88 non-null     Float64\n",
      " 43  aim_species_fraction_arab        88 non-null     float64\n",
      " 44  aim_species_fraction_colu        88 non-null     float64\n",
      " 45  aim_species_fraction_colu_no2l   88 non-null     float64\n",
      " 46  aim_species_gambcolu_arabiensis  88 non-null     object \n",
      " 47  aim_species_gambiae_coluzzii     88 non-null     object \n",
      " 48  aim_species                      88 non-null     object \n",
      " 49  country_iso                      88 non-null     object \n",
      " 50  admin1_name                      88 non-null     object \n",
      " 51  admin1_iso                       88 non-null     object \n",
      " 52  admin2_name                      88 non-null     object \n",
      " 53  taxon                            88 non-null     object \n",
      " 54  cohort_admin1_year               88 non-null     object \n",
      " 55  cohort_admin1_month              88 non-null     object \n",
      " 56  cohort_admin1_quarter            88 non-null     object \n",
      " 57  cohort_admin2_year               88 non-null     object \n",
      " 58  cohort_admin2_month              88 non-null     object \n",
      " 59  cohort_admin2_quarter            88 non-null     object \n",
      "dtypes: Float64(10), Int64(12), bool(1), float64(6), int64(3), object(28)\n",
      "memory usage: 43.2+ KB\n"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Loading phenotype data for '{demo_sample_set}' filtered by Deltamethrin ---\")\n",
    "df_deltamethrin = ag3.phenotype_data(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin'\"\n",
    ")\n",
    "\n",
    "print(f\"Shape of DataFrame: {df_deltamethrin.shape}\")\n",
    "print(\"\\nFirst 5 rows of the filtered DataFrame:\")\n",
    "df_deltamethrin.head()\n",
    "print(f\"\\nUnique insecticides in filtered data: {df_deltamethrin['insecticide'].unique()}\")\n",
    "print(\"\\nDataFrame Info:\")\n",
    "df_deltamethrin.info() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daede4ca-01cc-450a-8185-294033af85ad",
   "metadata": {},
   "source": [
    "### Example 2: Filter by multiple conditions (insecticide and dose)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eab7d81c-d80c-4279-9263-a0d7b425ff85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Loading phenotype data filtered by Deltamethrin and dose >= 1.0 ---\n",
      "Shape of DataFrame: (48, 60)\n",
      "\n",
      "First 5 rows of the multi-filtered DataFrame:\n",
      "\n",
      "Unique insecticides: ['Deltamethrin']\n",
      "Unique doses: [2.]\n",
      "\n",
      "DataFrame Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 48 entries, 40 to 87\n",
      "Data columns (total 60 columns):\n",
      " #   Column                           Non-Null Count  Dtype  \n",
      "---  ------                           --------------  -----  \n",
      " 0   sample_id                        48 non-null     object \n",
      " 1   country                          48 non-null     object \n",
      " 2   location                         48 non-null     object \n",
      " 3   insecticide                      48 non-null     object \n",
      " 4   dose                             48 non-null     float64\n",
      " 5   phenotype                        48 non-null     object \n",
      " 6   sample_set                       48 non-null     object \n",
      " 7   partner_sample_id                48 non-null     object \n",
      " 8   contributor                      48 non-null     object \n",
      " 9   year                             48 non-null     int64  \n",
      " 10  month                            48 non-null     int64  \n",
      " 11  latitude                         48 non-null     float64\n",
      " 12  longitude                        48 non-null     float64\n",
      " 13  sex_call                         48 non-null     object \n",
      " 14  release                          48 non-null     object \n",
      " 15  quarter                          48 non-null     int64  \n",
      " 16  study_id                         48 non-null     object \n",
      " 17  study_url                        48 non-null     object \n",
      " 18  terms_of_use_expiry_date         48 non-null     object \n",
      " 19  terms_of_use_url                 48 non-null     object \n",
      " 20  unrestricted_use                 48 non-null     bool   \n",
      " 21  mean_cov                         48 non-null     Float64\n",
      " 22  median_cov                       48 non-null     Int64  \n",
      " 23  modal_cov                        48 non-null     Int64  \n",
      " 24  mean_cov_2L                      48 non-null     Float64\n",
      " 25  median_cov_2L                    48 non-null     Int64  \n",
      " 26  mode_cov_2L                      48 non-null     Int64  \n",
      " 27  mean_cov_2R                      48 non-null     Float64\n",
      " 28  median_cov_2R                    48 non-null     Int64  \n",
      " 29  mode_cov_2R                      48 non-null     Int64  \n",
      " 30  mean_cov_3L                      48 non-null     Float64\n",
      " 31  median_cov_3L                    48 non-null     Int64  \n",
      " 32  mode_cov_3L                      48 non-null     Int64  \n",
      " 33  mean_cov_3R                      48 non-null     Float64\n",
      " 34  median_cov_3R                    48 non-null     Int64  \n",
      " 35  mode_cov_3R                      48 non-null     Int64  \n",
      " 36  mean_cov_X                       48 non-null     Float64\n",
      " 37  median_cov_X                     48 non-null     Int64  \n",
      " 38  mode_cov_X                       48 non-null     Int64  \n",
      " 39  frac_gen_cov                     48 non-null     Float64\n",
      " 40  divergence                       48 non-null     Float64\n",
      " 41  contam_pct                       48 non-null     Float64\n",
      " 42  contam_LLR                       48 non-null     Float64\n",
      " 43  aim_species_fraction_arab        48 non-null     float64\n",
      " 44  aim_species_fraction_colu        48 non-null     float64\n",
      " 45  aim_species_fraction_colu_no2l   48 non-null     float64\n",
      " 46  aim_species_gambcolu_arabiensis  48 non-null     object \n",
      " 47  aim_species_gambiae_coluzzii     48 non-null     object \n",
      " 48  aim_species                      48 non-null     object \n",
      " 49  country_iso                      48 non-null     object \n",
      " 50  admin1_name                      48 non-null     object \n",
      " 51  admin1_iso                       48 non-null     object \n",
      " 52  admin2_name                      48 non-null     object \n",
      " 53  taxon                            48 non-null     object \n",
      " 54  cohort_admin1_year               48 non-null     object \n",
      " 55  cohort_admin1_month              48 non-null     object \n",
      " 56  cohort_admin1_quarter            48 non-null     object \n",
      " 57  cohort_admin2_year               48 non-null     object \n",
      " 58  cohort_admin2_month              48 non-null     object \n",
      " 59  cohort_admin2_quarter            48 non-null     object \n",
      "dtypes: Float64(10), Int64(12), bool(1), float64(6), int64(3), object(28)\n",
      "memory usage: 23.6+ KB\n"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Loading phenotype data filtered by Deltamethrin and dose >= 1.0 ---\")\n",
    "df_filtered_multi = ag3.phenotype_data(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin' and dose >= 1.0\"\n",
    ")\n",
    "\n",
    "print(f\"Shape of DataFrame: {df_filtered_multi.shape}\")\n",
    "print(\"\\nFirst 5 rows of the multi-filtered DataFrame:\")\n",
    "df_filtered_multi.head()\n",
    "\n",
    "print(f\"\\nUnique insecticides: {df_filtered_multi['insecticide'].unique()}\")\n",
    "print(f\"Unique doses: {df_filtered_multi['dose'].unique()}\")\n",
    "print(\"\\nDataFrame Info:\")\n",
    "df_filtered_multi.info()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc1cc12c-b24b-41f3-b147-9e4d817ab588",
   "metadata": {},
   "source": [
    "### Example 3: Applying cohort size filtering\n",
    "Cohort filtering parameters (cohort_size, min_cohort_size, max_cohort_size) can be combined with sample_query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de8cd777-a4b2-44dd-a0bf-64a6ef419da7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Loading phenotype data with min_cohort_size=10 ---\n",
      "Shape of DataFrame: (88, 60)\n",
      "\n",
      "First 5 rows of the cohort-filtered DataFrame:\n",
      "\n",
      "DataFrame Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 88 entries, 0 to 87\n",
      "Data columns (total 60 columns):\n",
      " #   Column                           Non-Null Count  Dtype  \n",
      "---  ------                           --------------  -----  \n",
      " 0   sample_id                        88 non-null     object \n",
      " 1   country                          88 non-null     object \n",
      " 2   location                         88 non-null     object \n",
      " 3   insecticide                      88 non-null     object \n",
      " 4   dose                             88 non-null     float64\n",
      " 5   phenotype                        88 non-null     object \n",
      " 6   sample_set                       88 non-null     object \n",
      " 7   partner_sample_id                88 non-null     object \n",
      " 8   contributor                      88 non-null     object \n",
      " 9   year                             88 non-null     int64  \n",
      " 10  month                            88 non-null     int64  \n",
      " 11  latitude                         88 non-null     float64\n",
      " 12  longitude                        88 non-null     float64\n",
      " 13  sex_call                         88 non-null     object \n",
      " 14  release                          88 non-null     object \n",
      " 15  quarter                          88 non-null     int64  \n",
      " 16  study_id                         88 non-null     object \n",
      " 17  study_url                        88 non-null     object \n",
      " 18  terms_of_use_expiry_date         88 non-null     object \n",
      " 19  terms_of_use_url                 88 non-null     object \n",
      " 20  unrestricted_use                 88 non-null     bool   \n",
      " 21  mean_cov                         88 non-null     Float64\n",
      " 22  median_cov                       88 non-null     Int64  \n",
      " 23  modal_cov                        88 non-null     Int64  \n",
      " 24  mean_cov_2L                      88 non-null     Float64\n",
      " 25  median_cov_2L                    88 non-null     Int64  \n",
      " 26  mode_cov_2L                      88 non-null     Int64  \n",
      " 27  mean_cov_2R                      88 non-null     Float64\n",
      " 28  median_cov_2R                    88 non-null     Int64  \n",
      " 29  mode_cov_2R                      88 non-null     Int64  \n",
      " 30  mean_cov_3L                      88 non-null     Float64\n",
      " 31  median_cov_3L                    88 non-null     Int64  \n",
      " 32  mode_cov_3L                      88 non-null     Int64  \n",
      " 33  mean_cov_3R                      88 non-null     Float64\n",
      " 34  median_cov_3R                    88 non-null     Int64  \n",
      " 35  mode_cov_3R                      88 non-null     Int64  \n",
      " 36  mean_cov_X                       88 non-null     Float64\n",
      " 37  median_cov_X                     88 non-null     Int64  \n",
      " 38  mode_cov_X                       88 non-null     Int64  \n",
      " 39  frac_gen_cov                     88 non-null     Float64\n",
      " 40  divergence                       88 non-null     Float64\n",
      " 41  contam_pct                       88 non-null     Float64\n",
      " 42  contam_LLR                       88 non-null     Float64\n",
      " 43  aim_species_fraction_arab        88 non-null     float64\n",
      " 44  aim_species_fraction_colu        88 non-null     float64\n",
      " 45  aim_species_fraction_colu_no2l   88 non-null     float64\n",
      " 46  aim_species_gambcolu_arabiensis  88 non-null     object \n",
      " 47  aim_species_gambiae_coluzzii     88 non-null     object \n",
      " 48  aim_species                      88 non-null     object \n",
      " 49  country_iso                      88 non-null     object \n",
      " 50  admin1_name                      88 non-null     object \n",
      " 51  admin1_iso                       88 non-null     object \n",
      " 52  admin2_name                      88 non-null     object \n",
      " 53  taxon                            88 non-null     object \n",
      " 54  cohort_admin1_year               88 non-null     object \n",
      " 55  cohort_admin1_month              88 non-null     object \n",
      " 56  cohort_admin1_quarter            88 non-null     object \n",
      " 57  cohort_admin2_year               88 non-null     object \n",
      " 58  cohort_admin2_month              88 non-null     object \n",
      " 59  cohort_admin2_quarter            88 non-null     object \n",
      "dtypes: Float64(10), Int64(12), bool(1), float64(6), int64(3), object(28)\n",
      "memory usage: 42.7+ KB\n"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Loading phenotype data with min_cohort_size=10 ---\")\n",
    "df_cohort_filtered = ag3.phenotype_data(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin'\",\n",
    "    min_cohort_size=10\n",
    ")\n",
    "\n",
    "print(f\"Shape of DataFrame: {df_cohort_filtered.shape}\")\n",
    "print(\"\\nFirst 5 rows of the cohort-filtered DataFrame:\")\n",
    "df_cohort_filtered.head() \n",
    "print(\"\\nDataFrame Info:\")\n",
    "df_cohort_filtered.info() \n",
    "# Verify cohort sizes (optional, for internal testing)\n",
    "# if not df_cohort_filtered.empty:\n",
    "#     cohort_keys = [\"insecticide\", \"dose\", \"location\", \"country\", \"sample_set\"]\n",
    "#     available_keys = [col for col in cohort_keys if col in df_cohort_filtered.columns]\n",
    "#     if available_keys:\n",
    "#         cohort_sizes = df_cohort_filtered.groupby(available_keys).size()\n",
    "#         print(\"\\nCohort sizes after filtering:\")\n",
    "#         print(cohort_sizes)\n",
    "#         print(f\"All cohorts meet min_cohort_size (>=10): {all(cohort_sizes >= 10)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9a1f4cc9-528f-40bf-9efe-71aee41c21c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Getting binary phenotype outcomes for '1237-VO-BJ-DJOGBENOU-VMF00050' ---\n",
      "Shape of binary series: (88,)\n",
      "First 5 entries:\n",
      "sample_id\n",
      "VBS18949-5562STDY7801785    0.0\n",
      "VBS18950-5562STDY7801786    0.0\n",
      "VBS18951-5562STDY7801787    0.0\n",
      "VBS18952-5562STDY7801788    0.0\n",
      "VBS18953-5562STDY7801789    0.0\n",
      "Name: phenotype_binary, dtype: float64\n",
      "Unique values in series: [0. 1.]\n",
      "\n",
      "Shape of binary series (alive Deltamethrin): (48,)\n",
      "First 5 entries:\n",
      "sample_id\n",
      "VBS18995-5562STDY7801828    1.0\n",
      "VBS18996-5562STDY7801829    1.0\n",
      "VBS18998-5562STDY7801830    1.0\n",
      "VBS18999-5562STDY7801831    1.0\n",
      "VBS19000-5562STDY7801832    1.0\n",
      "Name: phenotype_binary, dtype: float64\n",
      "Unique values in series: [1.]\n",
      "\n",
      "Shape of binary series (dose 0.5): (40,)\n",
      "First 5 entries:\n",
      "sample_id\n",
      "VBS18949-5562STDY7801785    0.0\n",
      "VBS18950-5562STDY7801786    0.0\n",
      "VBS18951-5562STDY7801787    0.0\n",
      "VBS18952-5562STDY7801788    0.0\n",
      "VBS18953-5562STDY7801789    0.0\n",
      "Name: phenotype_binary, dtype: float64\n",
      "Unique values in series: [0.]\n"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Getting binary phenotype outcomes for '{demo_sample_set}' ---\")\n",
    "\n",
    "# Example 1: Binary outcomes for all Deltamethrin samples\n",
    "binary_deltamethrin = ag3.phenotype_binary(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin'\"\n",
    ")\n",
    "\n",
    "print(f\"Shape of binary series: {binary_deltamethrin.shape}\")\n",
    "print(\"First 5 entries:\")\n",
    "print(binary_deltamethrin.head())\n",
    "print(f\"Unique values in series: {binary_deltamethrin.unique()}\")\n",
    "\n",
    "\n",
    "# Example 2: Binary outcomes for samples that were 'alive' with Deltamethrin\n",
    "binary_alive_deltamethrin = ag3.phenotype_binary(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin' and phenotype == 'alive'\"\n",
    ")\n",
    "\n",
    "print(f\"\\nShape of binary series (alive Deltamethrin): {binary_alive_deltamethrin.shape}\")\n",
    "print(\"First 5 entries:\")\n",
    "print(binary_alive_deltamethrin.head())\n",
    "print(f\"Unique values in series: {binary_alive_deltamethrin.unique()}\")\n",
    "\n",
    "# Example 3: Binary outcomes for samples with dose 0.5\n",
    "binary_dose_0_5 = ag3.phenotype_binary(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"dose == 0.5\"\n",
    ")\n",
    "\n",
    "print(f\"\\nShape of binary series (dose 0.5): {binary_dose_0_5.shape}\")\n",
    "print(\"First 5 entries:\")\n",
    "print(binary_dose_0_5.head())\n",
    "print(f\"Unique values in series: {binary_dose_0_5.unique()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a92e430e-7f40-48a8-8c98-2405d27aabfb",
   "metadata": {},
   "source": [
    "## 4. Getting Binary Phenotype Outcomes with phenotype_binary\n",
    "\n",
    "The `phenotype_binary()` method provides a convenient way to get phenotype outcomes as a binary Pandas Series (1 for alive/resistant, 0 for dead/susceptible, NaN for unmapped). It also uses the `sample_query` for filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fda88368",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Getting binary phenotype outcomes for '1237-VO-BJ-DJOGBENOU-VMF00050' ---\n",
      "Shape of binary series: (88,)\n",
      "First 5 entries:\n",
      "sample_id\n",
      "VBS18949-5562STDY7801785    0.0\n",
      "VBS18950-5562STDY7801786    0.0\n",
      "VBS18951-5562STDY7801787    0.0\n",
      "VBS18952-5562STDY7801788    0.0\n",
      "VBS18953-5562STDY7801789    0.0\n",
      "Name: phenotype_binary, dtype: float64\n",
      "Unique values in series: [0. 1.]\n",
      "\n",
      "Shape of binary series (alive Deltamethrin): (48,)\n",
      "First 5 entries:\n",
      "sample_id\n",
      "VBS18995-5562STDY7801828    1.0\n",
      "VBS18996-5562STDY7801829    1.0\n",
      "VBS18998-5562STDY7801830    1.0\n",
      "VBS18999-5562STDY7801831    1.0\n",
      "VBS19000-5562STDY7801832    1.0\n",
      "Name: phenotype_binary, dtype: float64\n",
      "Unique values in series: [1.]\n",
      "\n",
      "Shape of binary series (dose 0.5): (40,)\n",
      "First 5 entries:\n",
      "sample_id\n",
      "VBS18949-5562STDY7801785    0.0\n",
      "VBS18950-5562STDY7801786    0.0\n",
      "VBS18951-5562STDY7801787    0.0\n",
      "VBS18952-5562STDY7801788    0.0\n",
      "VBS18953-5562STDY7801789    0.0\n",
      "Name: phenotype_binary, dtype: float64\n",
      "Unique values in series: [0.]\n"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Getting binary phenotype outcomes for '{demo_sample_set}' ---\")\n",
    "\n",
    "# Example 1: Binary outcomes for all Deltamethrin samples\n",
    "binary_deltamethrin = ag3.phenotype_binary(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin'\"\n",
    ")\n",
    "\n",
    "print(f\"Shape of binary series: {binary_deltamethrin.shape}\")\n",
    "print(\"First 5 entries:\")\n",
    "print(binary_deltamethrin.head())\n",
    "print(f\"Unique values in series: {binary_deltamethrin.unique()}\")\n",
    "\n",
    "\n",
    "# Example 2: Binary outcomes for samples that were 'alive' with Deltamethrin\n",
    "binary_alive_deltamethrin = ag3.phenotype_binary(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin' and phenotype == 'alive'\"\n",
    ")\n",
    "\n",
    "print(f\"\\nShape of binary series (alive Deltamethrin): {binary_alive_deltamethrin.shape}\")\n",
    "print(\"First 5 entries:\")\n",
    "print(binary_alive_deltamethrin.head())\n",
    "print(f\"Unique values in series: {binary_alive_deltamethrin.unique()}\")\n",
    "\n",
    "# Example 3: Binary outcomes for samples with dose 0.5\n",
    "binary_dose_0_5 = ag3.phenotype_binary(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"dose == 0.5\"\n",
    ")\n",
    "\n",
    "print(f\"\\nShape of binary series (dose 0.5): {binary_dose_0_5.shape}\")\n",
    "print(\"First 5 entries:\")\n",
    "print(binary_dose_0_5.head())\n",
    "print(f\"Unique values in series: {binary_dose_0_5.unique()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62e84ea0",
   "metadata": {},
   "source": [
    "## 5. Loading Phenotype Data Merged with SNP Calls\n",
    "The phenotypes_with_snps() method returns an xarray.Dataset that combines phenotype data with SNP calls for a specified genomic region. The sample_query parameter is still used to filter the phenotype data before merging."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "726ab8a3-6573-49d5-9b29-479176e5dee6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Loading phenotype data merged with SNP calls ---\n",
      "Dataset dimensions: FrozenMappingWarningOnValuesAccess({'samples': 48, 'variants': 10001, 'alleles': 4, 'ploidy': 2})\n",
      "\n",
      "Dataset structure (rich display):\n",
      "\n",
      "First 5 samples and their phenotype_binary values:\n",
      "[1. 1. 1. 1. 1.]\n",
      "\n",
      "First 5 samples and their insecticide values:\n",
      "['Deltamethrin' 'Deltamethrin' 'Deltamethrin' 'Deltamethrin'\n",
      " 'Deltamethrin']\n",
      "\n",
      "First 5 variant positions:\n",
      "[2420000 2420001 2420002 2420003 2420004]\n",
      "\n",
      "Dataset Info:\n",
      "xarray.Dataset {\n",
      "dimensions:\n",
      "\tsamples = 48 ;\n",
      "\tvariants = 10001 ;\n",
      "\talleles = 4 ;\n",
      "\tploidy = 2 ;\n",
      "\n",
      "variables:\n",
      "\tfloat64 phenotype_binary(samples) ;\n",
      "\tobject insecticide(samples) ;\n",
      "\tfloat64 dose(samples) ;\n",
      "\tobject phenotype(samples) ;\n",
      "\tobject location(samples) ;\n",
      "\tobject country(samples) ;\n",
      "\tobject sample_set(samples) ;\n",
      "\tobject samples(samples) ;\n",
      "\t|S1 variant_allele(variants, alleles) ;\n",
      "\tbool variant_filter_pass_gamb_colu_arab(variants) ;\n",
      "\tbool variant_filter_pass_gamb_colu(variants) ;\n",
      "\tbool variant_filter_pass_arab(variants) ;\n",
      "\tint32 variant_position(variants) ;\n",
      "\tuint8 variant_contig(variants) ;\n",
      "\tint8 call_genotype(variants, samples, ploidy) ;\n",
      "\tint16 call_GQ(variants, samples) ;\n",
      "\tfloat32 call_MQ(variants, samples) ;\n",
      "\tint16 call_AD(variants, samples, alleles) ;\n",
      "\tbool call_genotype_mask(variants, samples, ploidy) ;\n",
      "\n",
      "// global attributes:\n",
      "}"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Loading phenotype data merged with SNP calls ---\")\n",
    "\n",
    "# Choose a small region for faster loading\n",
    "demo_region_snps = \"2L:2420000-2430000\"\n",
    "\n",
    "ds_snps = ag3.phenotypes_with_snps(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin' and phenotype == 'alive'\",\n",
    "    region=demo_region_snps\n",
    ")\n",
    "\n",
    "print(f\"Dataset dimensions: {ds_snps.dims}\")\n",
    "print(\"\\nDataset structure (rich display):\")\n",
    "ds_snps\n",
    "\n",
    "print(\"\\nFirst 5 samples and their phenotype_binary values:\")\n",
    "print(ds_snps[\"phenotype_binary\"].head(5).values)\n",
    "print(\"\\nFirst 5 samples and their insecticide values:\")\n",
    "print(ds_snps[\"insecticide\"].head(5).values)\n",
    "print(\"\\nFirst 5 variant positions:\")\n",
    "print(ds_snps[\"variant_position\"].head(5).values)\n",
    "print(\"\\nDataset Info:\")\n",
    "ds_snps.info()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99e7b505-4ea4-43dc-91ae-14ac8f12076e",
   "metadata": {},
   "source": [
    "## 6. Loading Phenotype Data Merged with Haplotypes\n",
    "Similarly, the phenotypes_with_haplotypes() method returns an xarray.Dataset combining phenotype data with haplotype calls for a given region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "403f3a26-71eb-4781-975d-d94b7a021d97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Loading phenotype data merged with Haplotypes ---\n",
      "Dataset dimensions: FrozenMappingWarningOnValuesAccess({'samples': 40, 'variants': 2153, 'alleles': 2, 'ploidy': 2})\n",
      "\n",
      "Dataset structure (rich display):\n",
      "\n",
      "First 5 samples and their phenotype_binary values:\n",
      "[0. 0. 0. 0. 0.]\n",
      "\n",
      "First 5 samples and their insecticide values:\n",
      "['Deltamethrin' 'Deltamethrin' 'Deltamethrin' 'Deltamethrin'\n",
      " 'Deltamethrin']\n",
      "\n",
      "First 5 variant positions:\n",
      "[2420351 2420359 2420370 2420371 2420375]\n",
      "\n",
      "Dataset Info:\n",
      "xarray.Dataset {\n",
      "dimensions:\n",
      "\tsamples = 40 ;\n",
      "\tvariants = 2153 ;\n",
      "\talleles = 2 ;\n",
      "\tploidy = 2 ;\n",
      "\n",
      "variables:\n",
      "\tfloat64 phenotype_binary(samples) ;\n",
      "\tobject insecticide(samples) ;\n",
      "\tfloat64 dose(samples) ;\n",
      "\tobject phenotype(samples) ;\n",
      "\tobject location(samples) ;\n",
      "\tobject country(samples) ;\n",
      "\tobject sample_set(samples) ;\n",
      "\tobject samples(samples) ;\n",
      "\tint32 variant_position(variants) ;\n",
      "\tuint8 variant_contig(variants) ;\n",
      "\t|S1 variant_allele(variants, alleles) ;\n",
      "\tint8 call_genotype(variants, samples, ploidy) ;\n",
      "\n",
      "// global attributes:\n",
      "}"
     ]
    }
   ],
   "source": [
    "print(f\"\\n--- Loading phenotype data merged with Haplotypes ---\")\n",
    "\n",
    "# Choose a small region for faster loading\n",
    "demo_region_haps = \"2L:2420000-2430000\"\n",
    "\n",
    "ds_haps = ag3.phenotypes_with_haplotypes(\n",
    "    sample_sets=[demo_sample_set],\n",
    "    sample_query=\"insecticide == 'Deltamethrin' and phenotype == 'dead'\",\n",
    "    region=demo_region_haps\n",
    ")\n",
    "\n",
    "print(f\"Dataset dimensions: {ds_haps.dims}\")\n",
    "print(\"\\nDataset structure (rich display):\")\n",
    "ds_haps\n",
    "\n",
    "print(\"\\nFirst 5 samples and their phenotype_binary values:\")\n",
    "print(ds_haps[\"phenotype_binary\"].head(5).values)\n",
    "print(\"\\nFirst 5 samples and their insecticide values:\")\n",
    "print(ds_haps[\"insecticide\"].head(5).values)\n",
    "print(\"\\nFirst 5 variant positions:\")\n",
    "print(ds_haps[\"variant_position\"].head(5).values)\n",
    "print(\"\\nDataset Info:\")\n",
    "ds_haps.info()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ca345f-33e9-481d-8cf6-0ec08253121c",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "This notebook demonstrates the streamlined approach to accessing and combining phenotype and genetic data. The use of sample_query provides flexible filtering, and the dedicated functions for SNPs and Haplotypes (and potentially CNVs in the future) ensure clarity and separation of concerns."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
