{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore real data to help with simulating test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "import allel\n",
    "import numpy as np\n",
    "import plotly.express as px\n",
    "import plotly.io as pio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pio.templates.default = \"plotly_dark\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ag3 = malariagen_data.Ag3(\n",
    "    \"simplecache::gs://vo_agam_release_master_us_central1\",\n",
    "    simplecache=dict(cache_storage=\"../gcs_cache\"),\n",
    ")\n",
    "ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "af1 = malariagen_data.Af1(\n",
    "    \"simplecache::gs://vo_afun_release_master_us_central1\",\n",
    "    simplecache=dict(cache_storage=\"../gcs_cache\"),\n",
    ")\n",
    "af1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_snps = ag3.snp_calls(region=\"3L\", sample_sets=\"AG1000G-BF-A\")\n",
    "ds_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps = ag3.haplotypes(\n",
    "    region=\"3L\", sample_sets=\"AG1000G-BF-A\", analysis=\"gamb_colu_arab\"\n",
    ")\n",
    "ds_haps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps.sizes[\"variants\"] / ds_snps.sizes[\"variants\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps = ag3.haplotypes(region=\"3L\", sample_sets=\"AG1000G-BF-A\", analysis=\"gamb_colu\")\n",
    "ds_haps.sizes[\"variants\"] / ds_snps.sizes[\"variants\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps = ag3.haplotypes(region=\"3L\", sample_sets=\"AG1000G-UG\", analysis=\"arab\")\n",
    "ds_haps.sizes[\"variants\"] / ds_snps.sizes[\"variants\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_alleles = ds_snps[\"variant_allele\"].values\n",
    "snp_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_pos = ds_snps[\"variant_position\"].values\n",
    "snp_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loc_hap_sites = np.random.choice([False, True], size=snp_pos.shape[0], p=[0.75, 0.25])\n",
    "loc_hap_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_hap_sites = np.sum(loc_hap_sites)\n",
    "n_hap_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_hap_ref = snp_alleles[loc_hap_sites, 0]\n",
    "sim_hap_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.arange(1, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_alt_choice = np.random.choice(np.arange(1, 4), size=n_hap_sites)\n",
    "sim_alt_choice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_hap_alt = np.take_along_axis(\n",
    "    snp_alleles[loc_hap_sites], indices=sim_alt_choice[:, None], axis=1\n",
    ")[:, 0]\n",
    "sim_hap_alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_hap_alt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps = ag3.haplotypes(\n",
    "    region=\"3L\", sample_sets=\"AG1000G-BF-A\", analysis=\"gamb_colu_arab\"\n",
    ")\n",
    "gt = ds_haps[\"call_genotype\"][:1_000_000].values\n",
    "p_hap_01 = np.bincount(gt.flatten())\n",
    "p_hap_01 = p_hap_01 / p_hap_01.sum()\n",
    "p_hap_01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps = ag3.haplotypes(region=\"3L\", sample_sets=\"AG1000G-BF-A\", analysis=\"gamb_colu\")\n",
    "gt = ds_haps[\"call_genotype\"][:1_000_000].values\n",
    "p_hap_01 = np.bincount(gt.flatten())\n",
    "p_hap_01 = p_hap_01 / p_hap_01.sum()\n",
    "p_hap_01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_haps = ag3.haplotypes(region=\"3L\", sample_sets=\"AG1000G-UG\", analysis=\"arab\")\n",
    "gt = ds_haps[\"call_genotype\"][:1_000_000].values\n",
    "p_hap_01 = np.bincount(gt.flatten())\n",
    "p_hap_01 = p_hap_01 / p_hap_01.sum()\n",
    "p_hap_01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hap_gt_sim = np.random.choice(\n",
    "    [0, 1], size=(100_000, 100), replace=True, p=p_hap_01\n",
    ").astype(\"i1\")\n",
    "hap_gt_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(hap_gt_sim.flatten())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Site annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ann = ag3.open_site_annotations()\n",
    "ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ann.tree())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_degeneracy = ann[\"codon_degeneracy\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(codon_degeneracy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(codon_degeneracy + 1) / codon_degeneracy.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_nonsyn = ann[\"codon_nonsyn\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(codon_nonsyn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(codon_nonsyn) / codon_nonsyn.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_position = ann[\"codon_position\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(codon_position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(codon_position + 1) / codon_position.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_cls = ann[\"seq_cls\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(seq_cls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(seq_cls) / seq_cls.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_flen = ann[\"seq_flen\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(seq_flen, nbins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_relpos_start = ann[\"seq_relpos_start\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(seq_relpos_start, nbins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_relpos_stop = ann[\"seq_relpos_stop\"][\"3L\"][10_000_000:11_000_000]\n",
    "px.histogram(seq_relpos_stop, nbins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = np.random.beta(a=0.4, b=4, size=100_000) * 40_000\n",
    "px.histogram(x, nbins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(ann)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNP calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = ag3.snp_calls(region=\"3L\", sample_sets=\"AG1000G-BF-A\")\n",
    "ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds2 = ds.isel(variants=slice(10_500_000, 10_600_000))\n",
    "ds2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt = ds2[\"call_genotype\"].values\n",
    "gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt2 = allel.GenotypeArray(gt)\n",
    "gt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(gt.flatten() + 1) / gt.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_fraction = gt2.count_missing() / (gt2.n_variants * gt2.n_samples)\n",
    "missing_fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_counts = np.bincount(gt.flatten() + 1)[1:]\n",
    "allele_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_fractions = allele_counts / np.sum(allele_counts)\n",
    "allele_fractions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_sim = np.random.choice(4, size=gt.shape, replace=True, p=allele_fractions)\n",
    "gt_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_calls = gt_sim.shape[0] * gt_sim.shape[1]\n",
    "n_calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_sim_flat = gt_sim.reshape(-1, 2)\n",
    "gt_sim_flat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loc_missing = np.random.choice(\n",
    "    n_calls,\n",
    "    size=int(missing_fraction * n_calls),\n",
    "    replace=False,\n",
    ")\n",
    "loc_missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_sim_flat[loc_missing] = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(gt_sim.flatten() + 1) / gt_sim.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.bincount(gt.flatten() + 1) / gt.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.GenotypeArray(gt).count_missing()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.GenotypeArray(gt_sim).count_missing()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.GenotypeArray(gt).count_hom_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.GenotypeArray(gt_sim).count_hom_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.GenotypeArray(gt).count_hom_alt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.GenotypeArray(gt_sim).count_hom_alt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gq = ds2[\"call_GQ\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.histogram(gq.flatten()[:100_000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gqf = gq.flatten()\n",
    "n_gq = np.bincount(gqf[gqf >= 0])\n",
    "n_gq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_gq = np.bincount(gqf[gqf >= 0]) / gqf.size\n",
    "p_gq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.bar(p_gq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mq = ds2[\"call_MQ\"].values\n",
    "mq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mqf = mq.flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.histogram(mqf[:100_000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_mq = np.bincount(mqf[mqf >= 0]) / mqf.size\n",
    "p_mq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.bar(p_mq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad = ds2[\"call_AD\"].values\n",
    "ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad0 = ad[:, :, 0].flatten()\n",
    "ad0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad1 = ad[:, :, 1].flatten()\n",
    "ad1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad2 = ad[:, :, 2].flatten()\n",
    "ad3 = ad[:, :, 3].flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.histogram(ad0[:10_000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.histogram(ad1[:10_000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.histogram(ad2[:10_000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_ad0 = np.bincount(ad0[ad0 >= 0]) / ad0.size\n",
    "px.bar(p_ad0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_ad1 = np.bincount(ad1[ad1 >= 2]) / ad1.size\n",
    "px.bar(p_ad1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_ad2 = np.bincount(ad2[ad2 >= 2]) / ad1.size\n",
    "px.bar(p_ad2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_ad3 = np.bincount(ad3[ad3 >= 2]) / ad1.size\n",
    "px.bar(p_ad3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos = ds2[\"variant_position\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.line(pos[:100_000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alleles = ds2[\"variant_allele\"].values[:10]\n",
    "alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = alleles[:, 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_sim = np.empty(shape=(alleles.shape[0], alleles.shape[1] - 1), dtype=\"S1\")\n",
    "alt_sim[ref == b\"A\"] = np.array([b\"C\", b\"T\", b\"G\"])\n",
    "alt_sim[ref == b\"C\"] = np.array([b\"A\", b\"T\", b\"G\"])\n",
    "alt_sim[ref == b\"T\"] = np.array([b\"A\", b\"C\", b\"G\"])\n",
    "alt_sim[ref == b\"G\"] = np.array([b\"A\", b\"C\", b\"T\"])\n",
    "alt_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pass_gc = ds2[\"variant_filter_pass_gamb_colu\"].values\n",
    "p_pass_gc = np.sum(pass_gc) / pass_gc.size\n",
    "p_pass_gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pass_a = ds2[\"variant_filter_pass_arab\"].values\n",
    "p_pass_a = np.sum(pass_a) / pass_a.size\n",
    "p_pass_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pass_gca = ds2[\"variant_filter_pass_gamb_colu_arab\"].values\n",
    "p_pass_gca = np.sum(pass_gca) / pass_gca.size\n",
    "p_pass_gca"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ag_seq = ag3.genome_sequence(\"3L\").compute()\n",
    "ag_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ag_seq_count = Counter(ag_seq)\n",
    "ag_seq_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_pass = af1.site_filters(region=\"3RL:10,000,000-11,000,000\", mask=\"funestus\")\n",
    "filter_pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sum(filter_pass).compute() / filter_pass.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "af_seq = af1.genome_sequence(\"3RL\")[:60_000_000].compute()\n",
    "af_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bases = np.array([b\"a\", b\"c\", b\"g\", b\"t\", b\"n\", b\"A\", b\"C\", b\"G\", b\"T\", b\"N\"])\n",
    "bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "af_seq_count = Counter(af_seq)\n",
    "af_seq_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_bases_af = {b: af_seq_count[b] / af_seq.size for b in bases}\n",
    "p_bases_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_ad0 = np.bincount(ad0[ad0 >= 0]) / ad0.size\n",
    "px.bar(p_ad0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_bases_ag = {b: ag_seq_count[b] / ag_seq.size for b in bases}\n",
    "p_bases_ag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_bases_ag = np.array([ag_seq_count[b] for b in bases]) / ag_seq.size\n",
    "p_bases_ag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_bases_ag.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_sim = np.random.choice(bases, size=ag_seq.size, replace=True, p=p_bases_ag)\n",
    "seq_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "af1.sample_sets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = af1.snp_calls(\n",
    "    region=\"3RL:10_000_000-10_500_000\", sample_sets=\"1230-VO-GA-CF-AYALA-VMF00045\"\n",
    ")\n",
    "ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gt = ds[\"call_genotype\"].values\n",
    "gt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_counts = np.bincount(gt.flatten() + 1)[1:]\n",
    "allele_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_counts / np.sum(allele_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sum(gt < 0) / gt.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
